// @atproto/api/dist/client/types/com/atproto/label/defs.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'com.atproto.label.defs';
const hashLabel = 'label';
export function isLabel(v) {
    return is$typed(v, id, hashLabel);
}
export function validateLabel(v) {
    return validate(v, id, hashLabel);
}
const hashSelfLabels = 'selfLabels';
export function isSelfLabels(v) {
    return is$typed(v, id, hashSelfLabels);
}
export function validateSelfLabels(v) {
    return validate(v, id, hashSelfLabels);
}
const hashSelfLabel = 'selfLabel';
export function isSelfLabel(v) {
    return is$typed(v, id, hashSelfLabel);
}
export function validateSelfLabel(v) {
    return validate(v, id, hashSelfLabel);
}
const hashLabelValueDefinition = 'labelValueDefinition';
export function isLabelValueDefinition(v) {
    return is$typed(v, id, hashLabelValueDefinition);
}
export function validateLabelValueDefinition(v) {
    return validate(v, id, hashLabelValueDefinition);
}
const hashLabelValueDefinitionStrings = 'labelValueDefinitionStrings';
export function isLabelValueDefinitionStrings(v) {
    return is$typed(v, id, hashLabelValueDefinitionStrings);
}
export function validateLabelValueDefinitionStrings(v) {
    return validate(v, id, hashLabelValueDefinitionStrings);
}
