// @atproto/api/dist/client/types/app/bsky/notification/defs.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'app.bsky.notification.defs';
const hashRecordDeleted = 'recordDeleted';
export function isRecordDeleted(v) {
    return is$typed(v, id, hashRecordDeleted);
}
export function validateRecordDeleted(v) {
    return validate(v, id, hashRecordDeleted);
}
const hashChatPreference = 'chatPreference';
export function isChatPreference(v) {
    return is$typed(v, id, hashChatPreference);
}
export function validateChatPreference(v) {
    return validate(v, id, hashChatPreference);
}
const hashFilterablePreference = 'filterablePreference';
export function isFilterablePreference(v) {
    return is$typed(v, id, hashFilterablePreference);
}
export function validateFilterablePreference(v) {
    return validate(v, id, hashFilterablePreference);
}
const hashPreference = 'preference';
export function isPreference(v) {
    return is$typed(v, id, hashPreference);
}
export function validatePreference(v) {
    return validate(v, id, hashPreference);
}
const hashPreferences = 'preferences';
export function isPreferences(v) {
    return is$typed(v, id, hashPreferences);
}
export function validatePreferences(v) {
    return validate(v, id, hashPreferences);
}
const hashActivitySubscription = 'activitySubscription';
export function isActivitySubscription(v) {
    return is$typed(v, id, hashActivitySubscription);
}
export function validateActivitySubscription(v) {
    return validate(v, id, hashActivitySubscription);
}
const hashSubjectActivitySubscription = 'subjectActivitySubscription';
export function isSubjectActivitySubscription(v) {
    return is$typed(v, id, hashSubjectActivitySubscription);
}
export function validateSubjectActivitySubscription(v) {
    return validate(v, id, hashSubjectActivitySubscription);
}
