// @atproto/api/dist/client/types/com/atproto/repo/applyWrites.js
/**
 * GENERATED CODE - DO NOT MODIFY
 */
import { XRPCError } from '@atproto/xrpc';
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'com.atproto.repo.applyWrites';
export class InvalidSwapError extends XRPCError {
    constructor(src) {
        super(src.status, src.error, src.message, src.headers, { cause: src });
    }
}
export function toKnownErr(e) {
    if (e instanceof XRPCError) {
        if (e.error === 'InvalidSwap')
            return new InvalidSwapError(e);
    }
    return e;
}
const hashCreate = 'create';
export function isCreate(v) {
    return is$typed(v, id, hashCreate);
}
export function validateCreate(v) {
    return validate(v, id, hashCreate);
}
const hashUpdate = 'update';
export function isUpdate(v) {
    return is$typed(v, id, hashUpdate);
}
export function validateUpdate(v) {
    return validate(v, id, hashUpdate);
}
const hashDelete = 'delete';
export function isDelete(v) {
    return is$typed(v, id, hashDelete);
}
export function validateDelete(v) {
    return validate(v, id, hashDelete);
}
const hashCreateResult = 'createResult';
export function isCreateResult(v) {
    return is$typed(v, id, hashCreateResult);
}
export function validateCreateResult(v) {
    return validate(v, id, hashCreateResult);
}
const hashUpdateResult = 'updateResult';
export function isUpdateResult(v) {
    return is$typed(v, id, hashUpdateResult);
}
export function validateUpdateResult(v) {
    return validate(v, id, hashUpdateResult);
}
const hashDeleteResult = 'deleteResult';
export function isDeleteResult(v) {
    return is$typed(v, id, hashDeleteResult);
}
export function validateDeleteResult(v) {
    return validate(v, id, hashDeleteResult);
}
