// @atproto/api/dist/client/types/app/bsky/labeler/defs.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'app.bsky.labeler.defs';
const hashLabelerView = 'labelerView';
export function isLabelerView(v) {
    return is$typed(v, id, hashLabelerView);
}
export function validateLabelerView(v) {
    return validate(v, id, hashLabelerView);
}
const hashLabelerViewDetailed = 'labelerViewDetailed';
export function isLabelerViewDetailed(v) {
    return is$typed(v, id, hashLabelerViewDetailed);
}
export function validateLabelerViewDetailed(v) {
    return validate(v, id, hashLabelerViewDetailed);
}
const hashLabelerViewerState = 'labelerViewerState';
export function isLabelerViewerState(v) {
    return is$typed(v, id, hashLabelerViewerState);
}
export function validateLabelerViewerState(v) {
    return validate(v, id, hashLabelerViewerState);
}
const hashLabelerPolicies = 'labelerPolicies';
export function isLabelerPolicies(v) {
    return is$typed(v, id, hashLabelerPolicies);
}
export function validateLabelerPolicies(v) {
    return validate(v, id, hashLabelerPolicies);
}
