// @atproto/api/dist/client/types/app/bsky/graph/defs.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'app.bsky.graph.defs';
const hashListViewBasic = 'listViewBasic';
export function isListViewBasic(v) {
    return is$typed(v, id, hashListViewBasic);
}
export function validateListViewBasic(v) {
    return validate(v, id, hashListViewBasic);
}
const hashListView = 'listView';
export function isListView(v) {
    return is$typed(v, id, hashListView);
}
export function validateListView(v) {
    return validate(v, id, hashListView);
}
const hashListItemView = 'listItemView';
export function isListItemView(v) {
    return is$typed(v, id, hashListItemView);
}
export function validateListItemView(v) {
    return validate(v, id, hashListItemView);
}
const hashStarterPackView = 'starterPackView';
export function isStarterPackView(v) {
    return is$typed(v, id, hashStarterPackView);
}
export function validateStarterPackView(v) {
    return validate(v, id, hashStarterPackView);
}
const hashStarterPackViewBasic = 'starterPackViewBasic';
export function isStarterPackViewBasic(v) {
    return is$typed(v, id, hashStarterPackViewBasic);
}
export function validateStarterPackViewBasic(v) {
    return validate(v, id, hashStarterPackViewBasic);
}
/** A list of actors to apply an aggregate moderation action (mute/block) on. */
export const MODLIST = `${id}#modlist`;
/** A list of actors used for curation purposes such as list feeds or interaction gating. */
export const CURATELIST = `${id}#curatelist`;
/** A list of actors used for only for reference purposes such as within a starter pack. */
export const REFERENCELIST = `${id}#referencelist`;
const hashListViewerState = 'listViewerState';
export function isListViewerState(v) {
    return is$typed(v, id, hashListViewerState);
}
export function validateListViewerState(v) {
    return validate(v, id, hashListViewerState);
}
const hashNotFoundActor = 'notFoundActor';
export function isNotFoundActor(v) {
    return is$typed(v, id, hashNotFoundActor);
}
export function validateNotFoundActor(v) {
    return validate(v, id, hashNotFoundActor);
}
const hashRelationship = 'relationship';
export function isRelationship(v) {
    return is$typed(v, id, hashRelationship);
}
export function validateRelationship(v) {
    return validate(v, id, hashRelationship);
}
