// @atproto/api/dist/moderation/index.js
import { ModerationDecision } from './decision.js';
import { decideAccount } from './subjects/account.js';
import { decideFeedGenerator } from './subjects/feed-generator.js';
import { decideNotification } from './subjects/notification.js';
import { decidePost } from './subjects/post.js';
import { decideProfile } from './subjects/profile.js';
import { decideStatus } from './subjects/status.js';
import { decideUserList } from './subjects/user-list.js';
export { ModerationUI } from './ui.js';
export { ModerationDecision } from './decision.js';
export { hasMutedWord, matchMuteWords } from './mutewords.js';
export { interpretLabelValueDefinition, interpretLabelValueDefinitions, } from './util.js';
export function moderateProfile(subject, opts) {
    return ModerationDecision.merge(decideAccount(subject, opts), decideProfile(subject, opts));
}
export function moderatePost(subject, opts) {
    return decidePost(subject, opts);
}
export function moderateNotification(subject, opts) {
    return decideNotification(subject, opts);
}
export function moderateFeedGenerator(subject, opts) {
    return decideFeedGenerator(subject, opts);
}
export function moderateUserList(subject, opts) {
    return decideUserList(subject, opts);
}
export function moderateStatus(subject, opts) {
    return decideStatus(subject, opts);
}
