// @atproto/api/dist/client/types/app/bsky/unspecced/defs.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'app.bsky.unspecced.defs';
const hashSkeletonSearchPost = 'skeletonSearchPost';
export function isSkeletonSearchPost(v) {
    return is$typed(v, id, hashSkeletonSearchPost);
}
export function validateSkeletonSearchPost(v) {
    return validate(v, id, hashSkeletonSearchPost);
}
const hashSkeletonSearchActor = 'skeletonSearchActor';
export function isSkeletonSearchActor(v) {
    return is$typed(v, id, hashSkeletonSearchActor);
}
export function validateSkeletonSearchActor(v) {
    return validate(v, id, hashSkeletonSearchActor);
}
const hashSkeletonSearchStarterPack = 'skeletonSearchStarterPack';
export function isSkeletonSearchStarterPack(v) {
    return is$typed(v, id, hashSkeletonSearchStarterPack);
}
export function validateSkeletonSearchStarterPack(v) {
    return validate(v, id, hashSkeletonSearchStarterPack);
}
const hashTrendingTopic = 'trendingTopic';
export function isTrendingTopic(v) {
    return is$typed(v, id, hashTrendingTopic);
}
export function validateTrendingTopic(v) {
    return validate(v, id, hashTrendingTopic);
}
const hashSkeletonTrend = 'skeletonTrend';
export function isSkeletonTrend(v) {
    return is$typed(v, id, hashSkeletonTrend);
}
export function validateSkeletonTrend(v) {
    return validate(v, id, hashSkeletonTrend);
}
const hashTrendView = 'trendView';
export function isTrendView(v) {
    return is$typed(v, id, hashTrendView);
}
export function validateTrendView(v) {
    return validate(v, id, hashTrendView);
}
const hashThreadItemPost = 'threadItemPost';
export function isThreadItemPost(v) {
    return is$typed(v, id, hashThreadItemPost);
}
export function validateThreadItemPost(v) {
    return validate(v, id, hashThreadItemPost);
}
const hashThreadItemNoUnauthenticated = 'threadItemNoUnauthenticated';
export function isThreadItemNoUnauthenticated(v) {
    return is$typed(v, id, hashThreadItemNoUnauthenticated);
}
export function validateThreadItemNoUnauthenticated(v) {
    return validate(v, id, hashThreadItemNoUnauthenticated);
}
const hashThreadItemNotFound = 'threadItemNotFound';
export function isThreadItemNotFound(v) {
    return is$typed(v, id, hashThreadItemNotFound);
}
export function validateThreadItemNotFound(v) {
    return validate(v, id, hashThreadItemNotFound);
}
const hashThreadItemBlocked = 'threadItemBlocked';
export function isThreadItemBlocked(v) {
    return is$typed(v, id, hashThreadItemBlocked);
}
export function validateThreadItemBlocked(v) {
    return validate(v, id, hashThreadItemBlocked);
}
const hashAgeAssuranceState = 'ageAssuranceState';
export function isAgeAssuranceState(v) {
    return is$typed(v, id, hashAgeAssuranceState);
}
export function validateAgeAssuranceState(v) {
    return validate(v, id, hashAgeAssuranceState);
}
const hashAgeAssuranceEvent = 'ageAssuranceEvent';
export function isAgeAssuranceEvent(v) {
    return is$typed(v, id, hashAgeAssuranceEvent);
}
export function validateAgeAssuranceEvent(v) {
    return validate(v, id, hashAgeAssuranceEvent);
}
