// @atproto/api/dist/client/types/app/bsky/actor/defs.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'app.bsky.actor.defs';
const hashProfileViewBasic = 'profileViewBasic';
export function isProfileViewBasic(v) {
    return is$typed(v, id, hashProfileViewBasic);
}
export function validateProfileViewBasic(v) {
    return validate(v, id, hashProfileViewBasic);
}
const hashProfileView = 'profileView';
export function isProfileView(v) {
    return is$typed(v, id, hashProfileView);
}
export function validateProfileView(v) {
    return validate(v, id, hashProfileView);
}
const hashProfileViewDetailed = 'profileViewDetailed';
export function isProfileViewDetailed(v) {
    return is$typed(v, id, hashProfileViewDetailed);
}
export function validateProfileViewDetailed(v) {
    return validate(v, id, hashProfileViewDetailed);
}
const hashProfileAssociated = 'profileAssociated';
export function isProfileAssociated(v) {
    return is$typed(v, id, hashProfileAssociated);
}
export function validateProfileAssociated(v) {
    return validate(v, id, hashProfileAssociated);
}
const hashProfileAssociatedChat = 'profileAssociatedChat';
export function isProfileAssociatedChat(v) {
    return is$typed(v, id, hashProfileAssociatedChat);
}
export function validateProfileAssociatedChat(v) {
    return validate(v, id, hashProfileAssociatedChat);
}
const hashProfileAssociatedGerm = 'profileAssociatedGerm';
export function isProfileAssociatedGerm(v) {
    return is$typed(v, id, hashProfileAssociatedGerm);
}
export function validateProfileAssociatedGerm(v) {
    return validate(v, id, hashProfileAssociatedGerm);
}
const hashProfileAssociatedActivitySubscription = 'profileAssociatedActivitySubscription';
export function isProfileAssociatedActivitySubscription(v) {
    return is$typed(v, id, hashProfileAssociatedActivitySubscription);
}
export function validateProfileAssociatedActivitySubscription(v) {
    return validate(v, id, hashProfileAssociatedActivitySubscription);
}
const hashViewerState = 'viewerState';
export function isViewerState(v) {
    return is$typed(v, id, hashViewerState);
}
export function validateViewerState(v) {
    return validate(v, id, hashViewerState);
}
const hashKnownFollowers = 'knownFollowers';
export function isKnownFollowers(v) {
    return is$typed(v, id, hashKnownFollowers);
}
export function validateKnownFollowers(v) {
    return validate(v, id, hashKnownFollowers);
}
const hashVerificationState = 'verificationState';
export function isVerificationState(v) {
    return is$typed(v, id, hashVerificationState);
}
export function validateVerificationState(v) {
    return validate(v, id, hashVerificationState);
}
const hashVerificationView = 'verificationView';
export function isVerificationView(v) {
    return is$typed(v, id, hashVerificationView);
}
export function validateVerificationView(v) {
    return validate(v, id, hashVerificationView);
}
const hashAdultContentPref = 'adultContentPref';
export function isAdultContentPref(v) {
    return is$typed(v, id, hashAdultContentPref);
}
export function validateAdultContentPref(v) {
    return validate(v, id, hashAdultContentPref);
}
const hashContentLabelPref = 'contentLabelPref';
export function isContentLabelPref(v) {
    return is$typed(v, id, hashContentLabelPref);
}
export function validateContentLabelPref(v) {
    return validate(v, id, hashContentLabelPref);
}
const hashSavedFeed = 'savedFeed';
export function isSavedFeed(v) {
    return is$typed(v, id, hashSavedFeed);
}
export function validateSavedFeed(v) {
    return validate(v, id, hashSavedFeed);
}
const hashSavedFeedsPrefV2 = 'savedFeedsPrefV2';
export function isSavedFeedsPrefV2(v) {
    return is$typed(v, id, hashSavedFeedsPrefV2);
}
export function validateSavedFeedsPrefV2(v) {
    return validate(v, id, hashSavedFeedsPrefV2);
}
const hashSavedFeedsPref = 'savedFeedsPref';
export function isSavedFeedsPref(v) {
    return is$typed(v, id, hashSavedFeedsPref);
}
export function validateSavedFeedsPref(v) {
    return validate(v, id, hashSavedFeedsPref);
}
const hashPersonalDetailsPref = 'personalDetailsPref';
export function isPersonalDetailsPref(v) {
    return is$typed(v, id, hashPersonalDetailsPref);
}
export function validatePersonalDetailsPref(v) {
    return validate(v, id, hashPersonalDetailsPref);
}
const hashDeclaredAgePref = 'declaredAgePref';
export function isDeclaredAgePref(v) {
    return is$typed(v, id, hashDeclaredAgePref);
}
export function validateDeclaredAgePref(v) {
    return validate(v, id, hashDeclaredAgePref);
}
const hashFeedViewPref = 'feedViewPref';
export function isFeedViewPref(v) {
    return is$typed(v, id, hashFeedViewPref);
}
export function validateFeedViewPref(v) {
    return validate(v, id, hashFeedViewPref);
}
const hashThreadViewPref = 'threadViewPref';
export function isThreadViewPref(v) {
    return is$typed(v, id, hashThreadViewPref);
}
export function validateThreadViewPref(v) {
    return validate(v, id, hashThreadViewPref);
}
const hashInterestsPref = 'interestsPref';
export function isInterestsPref(v) {
    return is$typed(v, id, hashInterestsPref);
}
export function validateInterestsPref(v) {
    return validate(v, id, hashInterestsPref);
}
const hashMutedWord = 'mutedWord';
export function isMutedWord(v) {
    return is$typed(v, id, hashMutedWord);
}
export function validateMutedWord(v) {
    return validate(v, id, hashMutedWord);
}
const hashMutedWordsPref = 'mutedWordsPref';
export function isMutedWordsPref(v) {
    return is$typed(v, id, hashMutedWordsPref);
}
export function validateMutedWordsPref(v) {
    return validate(v, id, hashMutedWordsPref);
}
const hashHiddenPostsPref = 'hiddenPostsPref';
export function isHiddenPostsPref(v) {
    return is$typed(v, id, hashHiddenPostsPref);
}
export function validateHiddenPostsPref(v) {
    return validate(v, id, hashHiddenPostsPref);
}
const hashLabelersPref = 'labelersPref';
export function isLabelersPref(v) {
    return is$typed(v, id, hashLabelersPref);
}
export function validateLabelersPref(v) {
    return validate(v, id, hashLabelersPref);
}
const hashLabelerPrefItem = 'labelerPrefItem';
export function isLabelerPrefItem(v) {
    return is$typed(v, id, hashLabelerPrefItem);
}
export function validateLabelerPrefItem(v) {
    return validate(v, id, hashLabelerPrefItem);
}
const hashBskyAppStatePref = 'bskyAppStatePref';
export function isBskyAppStatePref(v) {
    return is$typed(v, id, hashBskyAppStatePref);
}
export function validateBskyAppStatePref(v) {
    return validate(v, id, hashBskyAppStatePref);
}
const hashBskyAppProgressGuide = 'bskyAppProgressGuide';
export function isBskyAppProgressGuide(v) {
    return is$typed(v, id, hashBskyAppProgressGuide);
}
export function validateBskyAppProgressGuide(v) {
    return validate(v, id, hashBskyAppProgressGuide);
}
const hashNux = 'nux';
export function isNux(v) {
    return is$typed(v, id, hashNux);
}
export function validateNux(v) {
    return validate(v, id, hashNux);
}
const hashVerificationPrefs = 'verificationPrefs';
export function isVerificationPrefs(v) {
    return is$typed(v, id, hashVerificationPrefs);
}
export function validateVerificationPrefs(v) {
    return validate(v, id, hashVerificationPrefs);
}
const hashLiveEventPreferences = 'liveEventPreferences';
export function isLiveEventPreferences(v) {
    return is$typed(v, id, hashLiveEventPreferences);
}
export function validateLiveEventPreferences(v) {
    return validate(v, id, hashLiveEventPreferences);
}
const hashPostInteractionSettingsPref = 'postInteractionSettingsPref';
export function isPostInteractionSettingsPref(v) {
    return is$typed(v, id, hashPostInteractionSettingsPref);
}
export function validatePostInteractionSettingsPref(v) {
    return validate(v, id, hashPostInteractionSettingsPref);
}
const hashStatusView = 'statusView';
export function isStatusView(v) {
    return is$typed(v, id, hashStatusView);
}
export function validateStatusView(v) {
    return validate(v, id, hashStatusView);
}
