// @atproto/api/dist/client/types/tools/ozone/moderation/defs.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'tools.ozone.moderation.defs';
const hashModEventView = 'modEventView';
export function isModEventView(v) {
    return is$typed(v, id, hashModEventView);
}
export function validateModEventView(v) {
    return validate(v, id, hashModEventView);
}
const hashModEventViewDetail = 'modEventViewDetail';
export function isModEventViewDetail(v) {
    return is$typed(v, id, hashModEventViewDetail);
}
export function validateModEventViewDetail(v) {
    return validate(v, id, hashModEventViewDetail);
}
const hashSubjectStatusView = 'subjectStatusView';
export function isSubjectStatusView(v) {
    return is$typed(v, id, hashSubjectStatusView);
}
export function validateSubjectStatusView(v) {
    return validate(v, id, hashSubjectStatusView);
}
const hashSubjectView = 'subjectView';
export function isSubjectView(v) {
    return is$typed(v, id, hashSubjectView);
}
export function validateSubjectView(v) {
    return validate(v, id, hashSubjectView);
}
const hashAccountStats = 'accountStats';
export function isAccountStats(v) {
    return is$typed(v, id, hashAccountStats);
}
export function validateAccountStats(v) {
    return validate(v, id, hashAccountStats);
}
const hashRecordsStats = 'recordsStats';
export function isRecordsStats(v) {
    return is$typed(v, id, hashRecordsStats);
}
export function validateRecordsStats(v) {
    return validate(v, id, hashRecordsStats);
}
const hashAccountStrike = 'accountStrike';
export function isAccountStrike(v) {
    return is$typed(v, id, hashAccountStrike);
}
export function validateAccountStrike(v) {
    return validate(v, id, hashAccountStrike);
}
/** Moderator review status of a subject: Open. Indicates that the subject needs to be reviewed by a moderator */
export const REVIEWOPEN = `${id}#reviewOpen`;
/** Moderator review status of a subject: Escalated. Indicates that the subject was escalated for review by a moderator */
export const REVIEWESCALATED = `${id}#reviewEscalated`;
/** Moderator review status of a subject: Closed. Indicates that the subject was already reviewed and resolved by a moderator */
export const REVIEWCLOSED = `${id}#reviewClosed`;
/** Moderator review status of a subject: Unnecessary. Indicates that the subject does not need a review at the moment but there is probably some moderation related metadata available for it */
export const REVIEWNONE = `${id}#reviewNone`;
const hashModEventTakedown = 'modEventTakedown';
export function isModEventTakedown(v) {
    return is$typed(v, id, hashModEventTakedown);
}
export function validateModEventTakedown(v) {
    return validate(v, id, hashModEventTakedown);
}
const hashModEventReverseTakedown = 'modEventReverseTakedown';
export function isModEventReverseTakedown(v) {
    return is$typed(v, id, hashModEventReverseTakedown);
}
export function validateModEventReverseTakedown(v) {
    return validate(v, id, hashModEventReverseTakedown);
}
const hashModEventResolveAppeal = 'modEventResolveAppeal';
export function isModEventResolveAppeal(v) {
    return is$typed(v, id, hashModEventResolveAppeal);
}
export function validateModEventResolveAppeal(v) {
    return validate(v, id, hashModEventResolveAppeal);
}
const hashModEventComment = 'modEventComment';
export function isModEventComment(v) {
    return is$typed(v, id, hashModEventComment);
}
export function validateModEventComment(v) {
    return validate(v, id, hashModEventComment);
}
const hashModEventReport = 'modEventReport';
export function isModEventReport(v) {
    return is$typed(v, id, hashModEventReport);
}
export function validateModEventReport(v) {
    return validate(v, id, hashModEventReport);
}
const hashModEventLabel = 'modEventLabel';
export function isModEventLabel(v) {
    return is$typed(v, id, hashModEventLabel);
}
export function validateModEventLabel(v) {
    return validate(v, id, hashModEventLabel);
}
const hashModEventPriorityScore = 'modEventPriorityScore';
export function isModEventPriorityScore(v) {
    return is$typed(v, id, hashModEventPriorityScore);
}
export function validateModEventPriorityScore(v) {
    return validate(v, id, hashModEventPriorityScore);
}
const hashAgeAssuranceEvent = 'ageAssuranceEvent';
export function isAgeAssuranceEvent(v) {
    return is$typed(v, id, hashAgeAssuranceEvent);
}
export function validateAgeAssuranceEvent(v) {
    return validate(v, id, hashAgeAssuranceEvent);
}
const hashAgeAssuranceOverrideEvent = 'ageAssuranceOverrideEvent';
export function isAgeAssuranceOverrideEvent(v) {
    return is$typed(v, id, hashAgeAssuranceOverrideEvent);
}
export function validateAgeAssuranceOverrideEvent(v) {
    return validate(v, id, hashAgeAssuranceOverrideEvent);
}
const hashAgeAssurancePurgeEvent = 'ageAssurancePurgeEvent';
export function isAgeAssurancePurgeEvent(v) {
    return is$typed(v, id, hashAgeAssurancePurgeEvent);
}
export function validateAgeAssurancePurgeEvent(v) {
    return validate(v, id, hashAgeAssurancePurgeEvent);
}
const hashRevokeAccountCredentialsEvent = 'revokeAccountCredentialsEvent';
export function isRevokeAccountCredentialsEvent(v) {
    return is$typed(v, id, hashRevokeAccountCredentialsEvent);
}
export function validateRevokeAccountCredentialsEvent(v) {
    return validate(v, id, hashRevokeAccountCredentialsEvent);
}
const hashModEventAcknowledge = 'modEventAcknowledge';
export function isModEventAcknowledge(v) {
    return is$typed(v, id, hashModEventAcknowledge);
}
export function validateModEventAcknowledge(v) {
    return validate(v, id, hashModEventAcknowledge);
}
const hashModEventEscalate = 'modEventEscalate';
export function isModEventEscalate(v) {
    return is$typed(v, id, hashModEventEscalate);
}
export function validateModEventEscalate(v) {
    return validate(v, id, hashModEventEscalate);
}
const hashModEventMute = 'modEventMute';
export function isModEventMute(v) {
    return is$typed(v, id, hashModEventMute);
}
export function validateModEventMute(v) {
    return validate(v, id, hashModEventMute);
}
const hashModEventUnmute = 'modEventUnmute';
export function isModEventUnmute(v) {
    return is$typed(v, id, hashModEventUnmute);
}
export function validateModEventUnmute(v) {
    return validate(v, id, hashModEventUnmute);
}
const hashModEventMuteReporter = 'modEventMuteReporter';
export function isModEventMuteReporter(v) {
    return is$typed(v, id, hashModEventMuteReporter);
}
export function validateModEventMuteReporter(v) {
    return validate(v, id, hashModEventMuteReporter);
}
const hashModEventUnmuteReporter = 'modEventUnmuteReporter';
export function isModEventUnmuteReporter(v) {
    return is$typed(v, id, hashModEventUnmuteReporter);
}
export function validateModEventUnmuteReporter(v) {
    return validate(v, id, hashModEventUnmuteReporter);
}
const hashModEventEmail = 'modEventEmail';
export function isModEventEmail(v) {
    return is$typed(v, id, hashModEventEmail);
}
export function validateModEventEmail(v) {
    return validate(v, id, hashModEventEmail);
}
const hashModEventDivert = 'modEventDivert';
export function isModEventDivert(v) {
    return is$typed(v, id, hashModEventDivert);
}
export function validateModEventDivert(v) {
    return validate(v, id, hashModEventDivert);
}
const hashModEventTag = 'modEventTag';
export function isModEventTag(v) {
    return is$typed(v, id, hashModEventTag);
}
export function validateModEventTag(v) {
    return validate(v, id, hashModEventTag);
}
const hashAccountEvent = 'accountEvent';
export function isAccountEvent(v) {
    return is$typed(v, id, hashAccountEvent);
}
export function validateAccountEvent(v) {
    return validate(v, id, hashAccountEvent);
}
const hashIdentityEvent = 'identityEvent';
export function isIdentityEvent(v) {
    return is$typed(v, id, hashIdentityEvent);
}
export function validateIdentityEvent(v) {
    return validate(v, id, hashIdentityEvent);
}
const hashRecordEvent = 'recordEvent';
export function isRecordEvent(v) {
    return is$typed(v, id, hashRecordEvent);
}
export function validateRecordEvent(v) {
    return validate(v, id, hashRecordEvent);
}
const hashScheduleTakedownEvent = 'scheduleTakedownEvent';
export function isScheduleTakedownEvent(v) {
    return is$typed(v, id, hashScheduleTakedownEvent);
}
export function validateScheduleTakedownEvent(v) {
    return validate(v, id, hashScheduleTakedownEvent);
}
const hashCancelScheduledTakedownEvent = 'cancelScheduledTakedownEvent';
export function isCancelScheduledTakedownEvent(v) {
    return is$typed(v, id, hashCancelScheduledTakedownEvent);
}
export function validateCancelScheduledTakedownEvent(v) {
    return validate(v, id, hashCancelScheduledTakedownEvent);
}
const hashRepoView = 'repoView';
export function isRepoView(v) {
    return is$typed(v, id, hashRepoView);
}
export function validateRepoView(v) {
    return validate(v, id, hashRepoView);
}
const hashRepoViewDetail = 'repoViewDetail';
export function isRepoViewDetail(v) {
    return is$typed(v, id, hashRepoViewDetail);
}
export function validateRepoViewDetail(v) {
    return validate(v, id, hashRepoViewDetail);
}
const hashRepoViewNotFound = 'repoViewNotFound';
export function isRepoViewNotFound(v) {
    return is$typed(v, id, hashRepoViewNotFound);
}
export function validateRepoViewNotFound(v) {
    return validate(v, id, hashRepoViewNotFound);
}
const hashRecordView = 'recordView';
export function isRecordView(v) {
    return is$typed(v, id, hashRecordView);
}
export function validateRecordView(v) {
    return validate(v, id, hashRecordView);
}
const hashRecordViewDetail = 'recordViewDetail';
export function isRecordViewDetail(v) {
    return is$typed(v, id, hashRecordViewDetail);
}
export function validateRecordViewDetail(v) {
    return validate(v, id, hashRecordViewDetail);
}
const hashRecordViewNotFound = 'recordViewNotFound';
export function isRecordViewNotFound(v) {
    return is$typed(v, id, hashRecordViewNotFound);
}
export function validateRecordViewNotFound(v) {
    return validate(v, id, hashRecordViewNotFound);
}
const hashConvoView = 'convoView';
export function isConvoView(v) {
    return is$typed(v, id, hashConvoView);
}
export function validateConvoView(v) {
    return validate(v, id, hashConvoView);
}
const hashModeration = 'moderation';
export function isModeration(v) {
    return is$typed(v, id, hashModeration);
}
export function validateModeration(v) {
    return validate(v, id, hashModeration);
}
const hashModerationDetail = 'moderationDetail';
export function isModerationDetail(v) {
    return is$typed(v, id, hashModerationDetail);
}
export function validateModerationDetail(v) {
    return validate(v, id, hashModerationDetail);
}
const hashBlobView = 'blobView';
export function isBlobView(v) {
    return is$typed(v, id, hashBlobView);
}
export function validateBlobView(v) {
    return validate(v, id, hashBlobView);
}
const hashImageDetails = 'imageDetails';
export function isImageDetails(v) {
    return is$typed(v, id, hashImageDetails);
}
export function validateImageDetails(v) {
    return validate(v, id, hashImageDetails);
}
const hashVideoDetails = 'videoDetails';
export function isVideoDetails(v) {
    return is$typed(v, id, hashVideoDetails);
}
export function validateVideoDetails(v) {
    return validate(v, id, hashVideoDetails);
}
const hashAccountHosting = 'accountHosting';
export function isAccountHosting(v) {
    return is$typed(v, id, hashAccountHosting);
}
export function validateAccountHosting(v) {
    return validate(v, id, hashAccountHosting);
}
const hashRecordHosting = 'recordHosting';
export function isRecordHosting(v) {
    return is$typed(v, id, hashRecordHosting);
}
export function validateRecordHosting(v) {
    return validate(v, id, hashRecordHosting);
}
const hashReporterStats = 'reporterStats';
export function isReporterStats(v) {
    return is$typed(v, id, hashReporterStats);
}
export function validateReporterStats(v) {
    return validate(v, id, hashReporterStats);
}
const hashModTool = 'modTool';
export function isModTool(v) {
    return is$typed(v, id, hashModTool);
}
export function validateModTool(v) {
    return validate(v, id, hashModTool);
}
/** Moderation event timeline event for a PLC create operation */
export const TIMELINEEVENTPLCCREATE = `${id}#timelineEventPlcCreate`;
/** Moderation event timeline event for generic PLC operation */
export const TIMELINEEVENTPLCOPERATION = `${id}#timelineEventPlcOperation`;
/** Moderation event timeline event for a PLC tombstone operation */
export const TIMELINEEVENTPLCTOMBSTONE = `${id}#timelineEventPlcTombstone`;
const hashScheduledActionView = 'scheduledActionView';
export function isScheduledActionView(v) {
    return is$typed(v, id, hashScheduledActionView);
}
export function validateScheduledActionView(v) {
    return validate(v, id, hashScheduledActionView);
}
