// @atproto/api/dist/client/types/tools/ozone/moderation/getAccountTimeline.js
/**
 * GENERATED CODE - DO NOT MODIFY
 */
import { XRPCError } from '@atproto/xrpc';
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'tools.ozone.moderation.getAccountTimeline';
export class RepoNotFoundError extends XRPCError {
    constructor(src) {
        super(src.status, src.error, src.message, src.headers, { cause: src });
    }
}
export function toKnownErr(e) {
    if (e instanceof XRPCError) {
        if (e.error === 'RepoNotFound')
            return new RepoNotFoundError(e);
    }
    return e;
}
const hashTimelineItem = 'timelineItem';
export function isTimelineItem(v) {
    return is$typed(v, id, hashTimelineItem);
}
export function validateTimelineItem(v) {
    return validate(v, id, hashTimelineItem);
}
const hashTimelineItemSummary = 'timelineItemSummary';
export function isTimelineItemSummary(v) {
    return is$typed(v, id, hashTimelineItemSummary);
}
export function validateTimelineItemSummary(v) {
    return validate(v, id, hashTimelineItemSummary);
}
