// @atproto/api/dist/client/types/tools/ozone/moderation/scheduleAction.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'tools.ozone.moderation.scheduleAction';
export function toKnownErr(e) {
    return e;
}
const hashTakedown = 'takedown';
export function isTakedown(v) {
    return is$typed(v, id, hashTakedown);
}
export function validateTakedown(v) {
    return validate(v, id, hashTakedown);
}
const hashSchedulingConfig = 'schedulingConfig';
export function isSchedulingConfig(v) {
    return is$typed(v, id, hashSchedulingConfig);
}
export function validateSchedulingConfig(v) {
    return validate(v, id, hashSchedulingConfig);
}
const hashScheduledActionResults = 'scheduledActionResults';
export function isScheduledActionResults(v) {
    return is$typed(v, id, hashScheduledActionResults);
}
export function validateScheduledActionResults(v) {
    return validate(v, id, hashScheduledActionResults);
}
const hashFailedScheduling = 'failedScheduling';
export function isFailedScheduling(v) {
    return is$typed(v, id, hashFailedScheduling);
}
export function validateFailedScheduling(v) {
    return validate(v, id, hashFailedScheduling);
}
