// @atproto/api/dist/client/types/app/bsky/draft/defs.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'app.bsky.draft.defs';
const hashDraftWithId = 'draftWithId';
export function isDraftWithId(v) {
    return is$typed(v, id, hashDraftWithId);
}
export function validateDraftWithId(v) {
    return validate(v, id, hashDraftWithId);
}
const hashDraft = 'draft';
export function isDraft(v) {
    return is$typed(v, id, hashDraft);
}
export function validateDraft(v) {
    return validate(v, id, hashDraft);
}
const hashDraftPost = 'draftPost';
export function isDraftPost(v) {
    return is$typed(v, id, hashDraftPost);
}
export function validateDraftPost(v) {
    return validate(v, id, hashDraftPost);
}
const hashDraftView = 'draftView';
export function isDraftView(v) {
    return is$typed(v, id, hashDraftView);
}
export function validateDraftView(v) {
    return validate(v, id, hashDraftView);
}
const hashDraftEmbedLocalRef = 'draftEmbedLocalRef';
export function isDraftEmbedLocalRef(v) {
    return is$typed(v, id, hashDraftEmbedLocalRef);
}
export function validateDraftEmbedLocalRef(v) {
    return validate(v, id, hashDraftEmbedLocalRef);
}
const hashDraftEmbedCaption = 'draftEmbedCaption';
export function isDraftEmbedCaption(v) {
    return is$typed(v, id, hashDraftEmbedCaption);
}
export function validateDraftEmbedCaption(v) {
    return validate(v, id, hashDraftEmbedCaption);
}
const hashDraftEmbedGallery = 'draftEmbedGallery';
export function isDraftEmbedGallery(v) {
    return is$typed(v, id, hashDraftEmbedGallery);
}
export function validateDraftEmbedGallery(v) {
    return validate(v, id, hashDraftEmbedGallery);
}
const hashDraftEmbedImage = 'draftEmbedImage';
export function isDraftEmbedImage(v) {
    return is$typed(v, id, hashDraftEmbedImage);
}
export function validateDraftEmbedImage(v) {
    return validate(v, id, hashDraftEmbedImage);
}
const hashDraftEmbedVideo = 'draftEmbedVideo';
export function isDraftEmbedVideo(v) {
    return is$typed(v, id, hashDraftEmbedVideo);
}
export function validateDraftEmbedVideo(v) {
    return validate(v, id, hashDraftEmbedVideo);
}
const hashDraftEmbedExternal = 'draftEmbedExternal';
export function isDraftEmbedExternal(v) {
    return is$typed(v, id, hashDraftEmbedExternal);
}
export function validateDraftEmbedExternal(v) {
    return validate(v, id, hashDraftEmbedExternal);
}
const hashDraftEmbedRecord = 'draftEmbedRecord';
export function isDraftEmbedRecord(v) {
    return is$typed(v, id, hashDraftEmbedRecord);
}
export function validateDraftEmbedRecord(v) {
    return validate(v, id, hashDraftEmbedRecord);
}
