// @atproto/api/dist/client/types/com/atproto/sync/subscribeRepos.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'com.atproto.sync.subscribeRepos';
const hashCommit = 'commit';
export function isCommit(v) {
    return is$typed(v, id, hashCommit);
}
export function validateCommit(v) {
    return validate(v, id, hashCommit);
}
const hashSync = 'sync';
export function isSync(v) {
    return is$typed(v, id, hashSync);
}
export function validateSync(v) {
    return validate(v, id, hashSync);
}
const hashIdentity = 'identity';
export function isIdentity(v) {
    return is$typed(v, id, hashIdentity);
}
export function validateIdentity(v) {
    return validate(v, id, hashIdentity);
}
const hashAccount = 'account';
export function isAccount(v) {
    return is$typed(v, id, hashAccount);
}
export function validateAccount(v) {
    return validate(v, id, hashAccount);
}
const hashInfo = 'info';
export function isInfo(v) {
    return is$typed(v, id, hashInfo);
}
export function validateInfo(v) {
    return validate(v, id, hashInfo);
}
const hashRepoOp = 'repoOp';
export function isRepoOp(v) {
    return is$typed(v, id, hashRepoOp);
}
export function validateRepoOp(v) {
    return validate(v, id, hashRepoOp);
}
