// @atproto/api/dist/client/types/app/bsky/ageassurance/defs.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'app.bsky.ageassurance.defs';
const hashState = 'state';
export function isState(v) {
    return is$typed(v, id, hashState);
}
export function validateState(v) {
    return validate(v, id, hashState);
}
const hashStateMetadata = 'stateMetadata';
export function isStateMetadata(v) {
    return is$typed(v, id, hashStateMetadata);
}
export function validateStateMetadata(v) {
    return validate(v, id, hashStateMetadata);
}
const hashConfig = 'config';
export function isConfig(v) {
    return is$typed(v, id, hashConfig);
}
export function validateConfig(v) {
    return validate(v, id, hashConfig);
}
const hashConfigRegion = 'configRegion';
export function isConfigRegion(v) {
    return is$typed(v, id, hashConfigRegion);
}
export function validateConfigRegion(v) {
    return validate(v, id, hashConfigRegion);
}
const hashConfigRegionRuleDefault = 'configRegionRuleDefault';
export function isConfigRegionRuleDefault(v) {
    return is$typed(v, id, hashConfigRegionRuleDefault);
}
export function validateConfigRegionRuleDefault(v) {
    return validate(v, id, hashConfigRegionRuleDefault);
}
const hashConfigRegionRuleIfDeclaredOverAge = 'configRegionRuleIfDeclaredOverAge';
export function isConfigRegionRuleIfDeclaredOverAge(v) {
    return is$typed(v, id, hashConfigRegionRuleIfDeclaredOverAge);
}
export function validateConfigRegionRuleIfDeclaredOverAge(v) {
    return validate(v, id, hashConfigRegionRuleIfDeclaredOverAge);
}
const hashConfigRegionRuleIfDeclaredUnderAge = 'configRegionRuleIfDeclaredUnderAge';
export function isConfigRegionRuleIfDeclaredUnderAge(v) {
    return is$typed(v, id, hashConfigRegionRuleIfDeclaredUnderAge);
}
export function validateConfigRegionRuleIfDeclaredUnderAge(v) {
    return validate(v, id, hashConfigRegionRuleIfDeclaredUnderAge);
}
const hashConfigRegionRuleIfAssuredOverAge = 'configRegionRuleIfAssuredOverAge';
export function isConfigRegionRuleIfAssuredOverAge(v) {
    return is$typed(v, id, hashConfigRegionRuleIfAssuredOverAge);
}
export function validateConfigRegionRuleIfAssuredOverAge(v) {
    return validate(v, id, hashConfigRegionRuleIfAssuredOverAge);
}
const hashConfigRegionRuleIfAssuredUnderAge = 'configRegionRuleIfAssuredUnderAge';
export function isConfigRegionRuleIfAssuredUnderAge(v) {
    return is$typed(v, id, hashConfigRegionRuleIfAssuredUnderAge);
}
export function validateConfigRegionRuleIfAssuredUnderAge(v) {
    return validate(v, id, hashConfigRegionRuleIfAssuredUnderAge);
}
const hashConfigRegionRuleIfAccountNewerThan = 'configRegionRuleIfAccountNewerThan';
export function isConfigRegionRuleIfAccountNewerThan(v) {
    return is$typed(v, id, hashConfigRegionRuleIfAccountNewerThan);
}
export function validateConfigRegionRuleIfAccountNewerThan(v) {
    return validate(v, id, hashConfigRegionRuleIfAccountNewerThan);
}
const hashConfigRegionRuleIfAccountOlderThan = 'configRegionRuleIfAccountOlderThan';
export function isConfigRegionRuleIfAccountOlderThan(v) {
    return is$typed(v, id, hashConfigRegionRuleIfAccountOlderThan);
}
export function validateConfigRegionRuleIfAccountOlderThan(v) {
    return validate(v, id, hashConfigRegionRuleIfAccountOlderThan);
}
const hashEvent = 'event';
export function isEvent(v) {
    return is$typed(v, id, hashEvent);
}
export function validateEvent(v) {
    return validate(v, id, hashEvent);
}
