// @atproto/api/dist/client/types/tools/ozone/report/defs.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'tools.ozone.report.defs';
/** Appeal a previously taken moderation action */
export const REASONAPPEAL = `${id}#reasonAppeal`;
/** An issue not included in these options */
export const REASONOTHER = `${id}#reasonOther`;
/** Animal welfare violations */
export const REASONVIOLENCEANIMAL = `${id}#reasonViolenceAnimal`;
/** Threats or incitement */
export const REASONVIOLENCETHREATS = `${id}#reasonViolenceThreats`;
/** Graphic violent content */
export const REASONVIOLENCEGRAPHICCONTENT = `${id}#reasonViolenceGraphicContent`;
/** Glorification of violence */
export const REASONVIOLENCEGLORIFICATION = `${id}#reasonViolenceGlorification`;
/** Extremist content. These reports will be sent only be sent to the application's Moderation Authority. */
export const REASONVIOLENCEEXTREMISTCONTENT = `${id}#reasonViolenceExtremistContent`;
/** Human trafficking */
export const REASONVIOLENCETRAFFICKING = `${id}#reasonViolenceTrafficking`;
/** Other violent content */
export const REASONVIOLENCEOTHER = `${id}#reasonViolenceOther`;
/** Adult sexual abuse content */
export const REASONSEXUALABUSECONTENT = `${id}#reasonSexualAbuseContent`;
/** Non-consensual intimate imagery */
export const REASONSEXUALNCII = `${id}#reasonSexualNCII`;
/** Deepfake adult content */
export const REASONSEXUALDEEPFAKE = `${id}#reasonSexualDeepfake`;
/** Animal sexual abuse */
export const REASONSEXUALANIMAL = `${id}#reasonSexualAnimal`;
/** Unlabelled adult content */
export const REASONSEXUALUNLABELED = `${id}#reasonSexualUnlabeled`;
/** Other sexual violence content */
export const REASONSEXUALOTHER = `${id}#reasonSexualOther`;
/** Child sexual abuse material (CSAM). These reports will be sent only be sent to the application's Moderation Authority. */
export const REASONCHILDSAFETYCSAM = `${id}#reasonChildSafetyCSAM`;
/** Grooming or predatory behavior. These reports will be sent only be sent to the application's Moderation Authority. */
export const REASONCHILDSAFETYGROOM = `${id}#reasonChildSafetyGroom`;
/** Privacy violation involving a minor */
export const REASONCHILDSAFETYPRIVACY = `${id}#reasonChildSafetyPrivacy`;
/** Harassment or bullying of minors */
export const REASONCHILDSAFETYHARASSMENT = `${id}#reasonChildSafetyHarassment`;
/** Other child safety. These reports will be sent only be sent to the application's Moderation Authority. */
export const REASONCHILDSAFETYOTHER = `${id}#reasonChildSafetyOther`;
/** Trolling */
export const REASONHARASSMENTTROLL = `${id}#reasonHarassmentTroll`;
/** Targeted harassment */
export const REASONHARASSMENTTARGETED = `${id}#reasonHarassmentTargeted`;
/** Hate speech */
export const REASONHARASSMENTHATESPEECH = `${id}#reasonHarassmentHateSpeech`;
/** Doxxing */
export const REASONHARASSMENTDOXXING = `${id}#reasonHarassmentDoxxing`;
/** Other harassing or hateful content */
export const REASONHARASSMENTOTHER = `${id}#reasonHarassmentOther`;
/** Fake account or bot */
export const REASONMISLEADINGBOT = `${id}#reasonMisleadingBot`;
/** Impersonation */
export const REASONMISLEADINGIMPERSONATION = `${id}#reasonMisleadingImpersonation`;
/** Spam */
export const REASONMISLEADINGSPAM = `${id}#reasonMisleadingSpam`;
/** Scam */
export const REASONMISLEADINGSCAM = `${id}#reasonMisleadingScam`;
/** False information about elections */
export const REASONMISLEADINGELECTIONS = `${id}#reasonMisleadingElections`;
/** Other misleading content */
export const REASONMISLEADINGOTHER = `${id}#reasonMisleadingOther`;
/** Hacking or system attacks */
export const REASONRULESITESECURITY = `${id}#reasonRuleSiteSecurity`;
/** Promoting or selling prohibited items or services */
export const REASONRULEPROHIBITEDSALES = `${id}#reasonRuleProhibitedSales`;
/** Banned user returning */
export const REASONRULEBANEVASION = `${id}#reasonRuleBanEvasion`;
/** Other */
export const REASONRULEOTHER = `${id}#reasonRuleOther`;
/** Content promoting or depicting self-harm */
export const REASONSELFHARMCONTENT = `${id}#reasonSelfHarmContent`;
/** Eating disorders */
export const REASONSELFHARMED = `${id}#reasonSelfHarmED`;
/** Dangerous challenges or activities */
export const REASONSELFHARMSTUNTS = `${id}#reasonSelfHarmStunts`;
/** Dangerous substances or drug abuse */
export const REASONSELFHARMSUBSTANCES = `${id}#reasonSelfHarmSubstances`;
/** Other dangerous content */
export const REASONSELFHARMOTHER = `${id}#reasonSelfHarmOther`;
const hashReportAssignment = 'reportAssignment';
export function isReportAssignment(v) {
    return is$typed(v, id, hashReportAssignment);
}
export function validateReportAssignment(v) {
    return validate(v, id, hashReportAssignment);
}
const hashReportView = 'reportView';
export function isReportView(v) {
    return is$typed(v, id, hashReportView);
}
export function validateReportView(v) {
    return validate(v, id, hashReportView);
}
const hashQueueActivity = 'queueActivity';
export function isQueueActivity(v) {
    return is$typed(v, id, hashQueueActivity);
}
export function validateQueueActivity(v) {
    return validate(v, id, hashQueueActivity);
}
const hashAssignmentActivity = 'assignmentActivity';
export function isAssignmentActivity(v) {
    return is$typed(v, id, hashAssignmentActivity);
}
export function validateAssignmentActivity(v) {
    return validate(v, id, hashAssignmentActivity);
}
const hashEscalationActivity = 'escalationActivity';
export function isEscalationActivity(v) {
    return is$typed(v, id, hashEscalationActivity);
}
export function validateEscalationActivity(v) {
    return validate(v, id, hashEscalationActivity);
}
const hashCloseActivity = 'closeActivity';
export function isCloseActivity(v) {
    return is$typed(v, id, hashCloseActivity);
}
export function validateCloseActivity(v) {
    return validate(v, id, hashCloseActivity);
}
const hashReopenActivity = 'reopenActivity';
export function isReopenActivity(v) {
    return is$typed(v, id, hashReopenActivity);
}
export function validateReopenActivity(v) {
    return validate(v, id, hashReopenActivity);
}
const hashNoteActivity = 'noteActivity';
export function isNoteActivity(v) {
    return is$typed(v, id, hashNoteActivity);
}
export function validateNoteActivity(v) {
    return validate(v, id, hashNoteActivity);
}
const hashReportActivityView = 'reportActivityView';
export function isReportActivityView(v) {
    return is$typed(v, id, hashReportActivityView);
}
export function validateReportActivityView(v) {
    return validate(v, id, hashReportActivityView);
}
const hashLiveStats = 'liveStats';
export function isLiveStats(v) {
    return is$typed(v, id, hashLiveStats);
}
export function validateLiveStats(v) {
    return validate(v, id, hashLiveStats);
}
const hashHistoricalStats = 'historicalStats';
export function isHistoricalStats(v) {
    return is$typed(v, id, hashHistoricalStats);
}
export function validateHistoricalStats(v) {
    return validate(v, id, hashHistoricalStats);
}
const hashAssignmentView = 'assignmentView';
export function isAssignmentView(v) {
    return is$typed(v, id, hashAssignmentView);
}
export function validateAssignmentView(v) {
    return validate(v, id, hashAssignmentView);
}
