// @atproto/api/dist/client/types/tools/ozone/report/reassignQueue.js
/**
 * GENERATED CODE - DO NOT MODIFY
 */
import { XRPCError } from '@atproto/xrpc';
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'tools.ozone.report.reassignQueue';
export class ReportNotFoundError extends XRPCError {
    constructor(src) {
        super(src.status, src.error, src.message, src.headers, { cause: src });
    }
}
export class ReportClosedError extends XRPCError {
    constructor(src) {
        super(src.status, src.error, src.message, src.headers, { cause: src });
    }
}
export class AlreadyInTargetQueueError extends XRPCError {
    constructor(src) {
        super(src.status, src.error, src.message, src.headers, { cause: src });
    }
}
export class QueueNotFoundError extends XRPCError {
    constructor(src) {
        super(src.status, src.error, src.message, src.headers, { cause: src });
    }
}
export class QueueDisabledError extends XRPCError {
    constructor(src) {
        super(src.status, src.error, src.message, src.headers, { cause: src });
    }
}
export function toKnownErr(e) {
    if (e instanceof XRPCError) {
        if (e.error === 'ReportNotFound')
            return new ReportNotFoundError(e);
        if (e.error === 'ReportClosed')
            return new ReportClosedError(e);
        if (e.error === 'AlreadyInTargetQueue')
            return new AlreadyInTargetQueueError(e);
        if (e.error === 'QueueNotFound')
            return new QueueNotFoundError(e);
        if (e.error === 'QueueDisabled')
            return new QueueDisabledError(e);
    }
    return e;
}
