// @atproto/api/dist/client/types/com/atproto/server/createAccount.js
/**
 * GENERATED CODE - DO NOT MODIFY
 */
import { XRPCError } from '@atproto/xrpc';
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'com.atproto.server.createAccount';
export class InvalidHandleError extends XRPCError {
    constructor(src) {
        super(src.status, src.error, src.message, src.headers, { cause: src });
    }
}
export class InvalidPasswordError extends XRPCError {
    constructor(src) {
        super(src.status, src.error, src.message, src.headers, { cause: src });
    }
}
export class InvalidInviteCodeError extends XRPCError {
    constructor(src) {
        super(src.status, src.error, src.message, src.headers, { cause: src });
    }
}
export class HandleNotAvailableError extends XRPCError {
    constructor(src) {
        super(src.status, src.error, src.message, src.headers, { cause: src });
    }
}
export class UnsupportedDomainError extends XRPCError {
    constructor(src) {
        super(src.status, src.error, src.message, src.headers, { cause: src });
    }
}
export class UnresolvableDidError extends XRPCError {
    constructor(src) {
        super(src.status, src.error, src.message, src.headers, { cause: src });
    }
}
export class IncompatibleDidDocError extends XRPCError {
    constructor(src) {
        super(src.status, src.error, src.message, src.headers, { cause: src });
    }
}
export function toKnownErr(e) {
    if (e instanceof XRPCError) {
        if (e.error === 'InvalidHandle')
            return new InvalidHandleError(e);
        if (e.error === 'InvalidPassword')
            return new InvalidPasswordError(e);
        if (e.error === 'InvalidInviteCode')
            return new InvalidInviteCodeError(e);
        if (e.error === 'HandleNotAvailable')
            return new HandleNotAvailableError(e);
        if (e.error === 'UnsupportedDomain')
            return new UnsupportedDomainError(e);
        if (e.error === 'UnresolvableDid')
            return new UnresolvableDidError(e);
        if (e.error === 'IncompatibleDidDoc')
            return new IncompatibleDidDocError(e);
    }
    return e;
}
