// @atproto/api/dist/client/types/tools/ozone/hosting/getAccountHistory.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'tools.ozone.hosting.getAccountHistory';
export function toKnownErr(e) {
    return e;
}
const hashEvent = 'event';
export function isEvent(v) {
    return is$typed(v, id, hashEvent);
}
export function validateEvent(v) {
    return validate(v, id, hashEvent);
}
const hashAccountCreated = 'accountCreated';
export function isAccountCreated(v) {
    return is$typed(v, id, hashAccountCreated);
}
export function validateAccountCreated(v) {
    return validate(v, id, hashAccountCreated);
}
const hashEmailUpdated = 'emailUpdated';
export function isEmailUpdated(v) {
    return is$typed(v, id, hashEmailUpdated);
}
export function validateEmailUpdated(v) {
    return validate(v, id, hashEmailUpdated);
}
const hashEmailConfirmed = 'emailConfirmed';
export function isEmailConfirmed(v) {
    return is$typed(v, id, hashEmailConfirmed);
}
export function validateEmailConfirmed(v) {
    return validate(v, id, hashEmailConfirmed);
}
const hashPasswordUpdated = 'passwordUpdated';
export function isPasswordUpdated(v) {
    return is$typed(v, id, hashPasswordUpdated);
}
export function validatePasswordUpdated(v) {
    return validate(v, id, hashPasswordUpdated);
}
const hashHandleUpdated = 'handleUpdated';
export function isHandleUpdated(v) {
    return is$typed(v, id, hashHandleUpdated);
}
export function validateHandleUpdated(v) {
    return validate(v, id, hashHandleUpdated);
}
