// unsplash-js/dist/index.mjs
import { trackNonHotLinkedPhotoView } from "./beacon.mjs";
import createFetchClient from "openapi-fetch";
//#region src/index.ts
const DEFAULT_BASE_URL = "https://api.unsplash.com/";
const API_VERSION = "v1";
/**
* Creates an {@link https://openapi-ts.dev/openapi-fetch/|openapi-fetch} client for the unsplash api.
*
* @see {@link https://unsplash.com/documentation|Unsplash's documentation} for endpoints details.
* @see {@link https://openapi-ts.dev/openapi-fetch/|openapi-fetch's documentation} for advanced usage.
*
* @example
* ```
* // Simple initialization
* import { createClient } from "unsplash-js";
* const client = createApi({
*   accessKey: process.env.UNSPLASH_ACCESS_KEY
* });
*
* await client.GET("/photos/{assetSlug}", {
*   params: {
*     path: { assetSlug: "GuRkV6x5L8L" }
*   }
* })
* ```
*/
const createApi = ({ accessKey, apiVersion = API_VERSION, headers, baseUrl = DEFAULT_BASE_URL, ...config } = {}) => {
	return createFetchClient({
		...config,
		querySerializer: {
			array: {
				style: "form",
				explode: false
			},
			...config.querySerializer
		},
		baseUrl,
		headers: {
			...headers,
			"Accept-Version": apiVersion,
			...accessKey !== void 0 ? { Authorization: `Client-ID ${accessKey}` } : {}
		}
	});
};
//#endregion
export { createApi, trackNonHotLinkedPhotoView };
