// unsplash-js/dist/index.cjs
Object.defineProperty(exports, Symbol.toStringTag, { value: "Module" });
const require_runtime = require("./_virtual/_rolldown/runtime.cjs");
const require_beacon = require("./beacon.cjs");
let openapi_fetch = require("openapi-fetch");
openapi_fetch = require_runtime.__toESM(openapi_fetch);
//#region src/index.ts
const DEFAULT_BASE_URL = "https://api.unsplash.com/";
const API_VERSION = "v1";
/**
* Creates an {@link https://openapi-ts.dev/openapi-fetch/|openapi-fetch} client for the unsplash api.
*
* @see {@link https://unsplash.com/documentation|Unsplash's documentation} for endpoints details.
* @see {@link https://openapi-ts.dev/openapi-fetch/|openapi-fetch's documentation} for advanced usage.
*
* @example
* ```
* // Simple initialization
* import { createClient } from "unsplash-js";
* const client = createApi({
*   accessKey: process.env.UNSPLASH_ACCESS_KEY
* });
*
* await client.GET("/photos/{assetSlug}", {
*   params: {
*     path: { assetSlug: "GuRkV6x5L8L" }
*   }
* })
* ```
*/
const createApi = ({ accessKey, apiVersion = API_VERSION, headers, baseUrl = DEFAULT_BASE_URL, ...config } = {}) => {
	return (0, openapi_fetch.default)({
		...config,
		querySerializer: {
			array: {
				style: "form",
				explode: false
			},
			...config.querySerializer
		},
		baseUrl,
		headers: {
			...headers,
			"Accept-Version": apiVersion,
			...accessKey !== void 0 ? { Authorization: `Client-ID ${accessKey}` } : {}
		}
	});
};
//#endregion
exports.createApi = createApi;
exports.trackNonHotLinkedPhotoView = require_beacon.trackNonHotLinkedPhotoView;
