// @atproto/xrpc/dist/client.js
import { Lexicons } from '@atproto/lexicon';
import { combineHeaders } from './util.js';
import { XrpcClient } from './xrpc-client.js';
/** @deprecated Use {@link XrpcClient} instead */
export class Client {
    constructor() {
        this.lex = new Lexicons();
    }
    /** @deprecated */
    get fetch() {
        throw new Error('Client.fetch is no longer supported. Use an XrpcClient instead.');
    }
    /** @deprecated */
    set fetch(_) {
        throw new Error('Client.fetch is no longer supported. Use an XrpcClient instead.');
    }
    // method calls
    //
    async call(serviceUri, methodNsid, params, data, opts) {
        return this.service(serviceUri).call(methodNsid, params, data, opts);
    }
    service(serviceUri) {
        return new ServiceClient(this, serviceUri);
    }
    // schemas
    // =
    addLexicon(doc) {
        this.lex.add(doc);
    }
    addLexicons(docs) {
        for (const doc of docs) {
            this.addLexicon(doc);
        }
    }
    removeLexicon(uri) {
        this.lex.remove(uri);
    }
}
/** @deprecated Use {@link XrpcClient} instead */
export class ServiceClient extends XrpcClient {
    constructor(baseClient, serviceUri) {
        super(async (input, init) => {
            const headers = combineHeaders(init.headers, Object.entries(this.headers));
            return fetch(new URL(input, this.uri), { ...init, headers });
        }, baseClient.lex);
        this.baseClient = baseClient;
        this.uri = typeof serviceUri === 'string' ? new URL(serviceUri) : serviceUri;
    }
}
