// @atproto/api/dist/client/types/app/bsky/richtext/facet.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'app.bsky.richtext.facet';
const hashMain = 'main';
export function isMain(v) {
    return is$typed(v, id, hashMain);
}
export function validateMain(v) {
    return validate(v, id, hashMain);
}
const hashMention = 'mention';
export function isMention(v) {
    return is$typed(v, id, hashMention);
}
export function validateMention(v) {
    return validate(v, id, hashMention);
}
const hashLink = 'link';
export function isLink(v) {
    return is$typed(v, id, hashLink);
}
export function validateLink(v) {
    return validate(v, id, hashLink);
}
const hashTag = 'tag';
export function isTag(v) {
    return is$typed(v, id, hashTag);
}
export function validateTag(v) {
    return validate(v, id, hashTag);
}
const hashByteSlice = 'byteSlice';
export function isByteSlice(v) {
    return is$typed(v, id, hashByteSlice);
}
export function validateByteSlice(v) {
    return validate(v, id, hashByteSlice);
}
