// @atproto/api/dist/client/types/app/bsky/embed/external.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'app.bsky.embed.external';
const hashMain = 'main';
export function isMain(v) {
    return is$typed(v, id, hashMain);
}
export function validateMain(v) {
    return validate(v, id, hashMain);
}
const hashExternal = 'external';
export function isExternal(v) {
    return is$typed(v, id, hashExternal);
}
export function validateExternal(v) {
    return validate(v, id, hashExternal);
}
const hashView = 'view';
export function isView(v) {
    return is$typed(v, id, hashView);
}
export function validateView(v) {
    return validate(v, id, hashView);
}
const hashViewExternal = 'viewExternal';
export function isViewExternal(v) {
    return is$typed(v, id, hashViewExternal);
}
export function validateViewExternal(v) {
    return validate(v, id, hashViewExternal);
}
const hashViewExternalSource = 'viewExternalSource';
export function isViewExternalSource(v) {
    return is$typed(v, id, hashViewExternalSource);
}
export function validateViewExternalSource(v) {
    return validate(v, id, hashViewExternalSource);
}
const hashViewExternalSourceTheme = 'viewExternalSourceTheme';
export function isViewExternalSourceTheme(v) {
    return is$typed(v, id, hashViewExternalSourceTheme);
}
export function validateViewExternalSourceTheme(v) {
    return validate(v, id, hashViewExternalSourceTheme);
}
const hashColorRGB = 'colorRGB';
export function isColorRGB(v) {
    return is$typed(v, id, hashColorRGB);
}
export function validateColorRGB(v) {
    return validate(v, id, hashColorRGB);
}
