// @atproto/api/dist/client/types/app/bsky/embed/record.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'app.bsky.embed.record';
const hashMain = 'main';
export function isMain(v) {
    return is$typed(v, id, hashMain);
}
export function validateMain(v) {
    return validate(v, id, hashMain);
}
const hashView = 'view';
export function isView(v) {
    return is$typed(v, id, hashView);
}
export function validateView(v) {
    return validate(v, id, hashView);
}
const hashViewRecord = 'viewRecord';
export function isViewRecord(v) {
    return is$typed(v, id, hashViewRecord);
}
export function validateViewRecord(v) {
    return validate(v, id, hashViewRecord);
}
const hashViewNotFound = 'viewNotFound';
export function isViewNotFound(v) {
    return is$typed(v, id, hashViewNotFound);
}
export function validateViewNotFound(v) {
    return validate(v, id, hashViewNotFound);
}
const hashViewBlocked = 'viewBlocked';
export function isViewBlocked(v) {
    return is$typed(v, id, hashViewBlocked);
}
export function validateViewBlocked(v) {
    return validate(v, id, hashViewBlocked);
}
const hashViewDetached = 'viewDetached';
export function isViewDetached(v) {
    return is$typed(v, id, hashViewDetached);
}
export function validateViewDetached(v) {
    return validate(v, id, hashViewDetached);
}
