// @atproto/api/dist/client/types/app/bsky/feed/defs.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'app.bsky.feed.defs';
const hashPostView = 'postView';
export function isPostView(v) {
    return is$typed(v, id, hashPostView);
}
export function validatePostView(v) {
    return validate(v, id, hashPostView);
}
const hashViewerState = 'viewerState';
export function isViewerState(v) {
    return is$typed(v, id, hashViewerState);
}
export function validateViewerState(v) {
    return validate(v, id, hashViewerState);
}
const hashThreadContext = 'threadContext';
export function isThreadContext(v) {
    return is$typed(v, id, hashThreadContext);
}
export function validateThreadContext(v) {
    return validate(v, id, hashThreadContext);
}
const hashFeedViewPost = 'feedViewPost';
export function isFeedViewPost(v) {
    return is$typed(v, id, hashFeedViewPost);
}
export function validateFeedViewPost(v) {
    return validate(v, id, hashFeedViewPost);
}
const hashReplyRef = 'replyRef';
export function isReplyRef(v) {
    return is$typed(v, id, hashReplyRef);
}
export function validateReplyRef(v) {
    return validate(v, id, hashReplyRef);
}
const hashReasonRepost = 'reasonRepost';
export function isReasonRepost(v) {
    return is$typed(v, id, hashReasonRepost);
}
export function validateReasonRepost(v) {
    return validate(v, id, hashReasonRepost);
}
const hashReasonPin = 'reasonPin';
export function isReasonPin(v) {
    return is$typed(v, id, hashReasonPin);
}
export function validateReasonPin(v) {
    return validate(v, id, hashReasonPin);
}
const hashThreadViewPost = 'threadViewPost';
export function isThreadViewPost(v) {
    return is$typed(v, id, hashThreadViewPost);
}
export function validateThreadViewPost(v) {
    return validate(v, id, hashThreadViewPost);
}
const hashNotFoundPost = 'notFoundPost';
export function isNotFoundPost(v) {
    return is$typed(v, id, hashNotFoundPost);
}
export function validateNotFoundPost(v) {
    return validate(v, id, hashNotFoundPost);
}
const hashBlockedPost = 'blockedPost';
export function isBlockedPost(v) {
    return is$typed(v, id, hashBlockedPost);
}
export function validateBlockedPost(v) {
    return validate(v, id, hashBlockedPost);
}
const hashBlockedAuthor = 'blockedAuthor';
export function isBlockedAuthor(v) {
    return is$typed(v, id, hashBlockedAuthor);
}
export function validateBlockedAuthor(v) {
    return validate(v, id, hashBlockedAuthor);
}
const hashGeneratorView = 'generatorView';
export function isGeneratorView(v) {
    return is$typed(v, id, hashGeneratorView);
}
export function validateGeneratorView(v) {
    return validate(v, id, hashGeneratorView);
}
const hashGeneratorViewerState = 'generatorViewerState';
export function isGeneratorViewerState(v) {
    return is$typed(v, id, hashGeneratorViewerState);
}
export function validateGeneratorViewerState(v) {
    return validate(v, id, hashGeneratorViewerState);
}
const hashSkeletonFeedPost = 'skeletonFeedPost';
export function isSkeletonFeedPost(v) {
    return is$typed(v, id, hashSkeletonFeedPost);
}
export function validateSkeletonFeedPost(v) {
    return validate(v, id, hashSkeletonFeedPost);
}
const hashSkeletonReasonRepost = 'skeletonReasonRepost';
export function isSkeletonReasonRepost(v) {
    return is$typed(v, id, hashSkeletonReasonRepost);
}
export function validateSkeletonReasonRepost(v) {
    return validate(v, id, hashSkeletonReasonRepost);
}
const hashSkeletonReasonPin = 'skeletonReasonPin';
export function isSkeletonReasonPin(v) {
    return is$typed(v, id, hashSkeletonReasonPin);
}
export function validateSkeletonReasonPin(v) {
    return validate(v, id, hashSkeletonReasonPin);
}
const hashThreadgateView = 'threadgateView';
export function isThreadgateView(v) {
    return is$typed(v, id, hashThreadgateView);
}
export function validateThreadgateView(v) {
    return validate(v, id, hashThreadgateView);
}
const hashInteraction = 'interaction';
export function isInteraction(v) {
    return is$typed(v, id, hashInteraction);
}
export function validateInteraction(v) {
    return validate(v, id, hashInteraction);
}
/** Request that less content like the given feed item be shown in the feed */
export const REQUESTLESS = `${id}#requestLess`;
/** Request that more content like the given feed item be shown in the feed */
export const REQUESTMORE = `${id}#requestMore`;
/** User clicked through to the feed item */
export const CLICKTHROUGHITEM = `${id}#clickthroughItem`;
/** User clicked through to the author of the feed item */
export const CLICKTHROUGHAUTHOR = `${id}#clickthroughAuthor`;
/** User clicked through to the reposter of the feed item */
export const CLICKTHROUGHREPOSTER = `${id}#clickthroughReposter`;
/** User clicked through to the embedded content of the feed item */
export const CLICKTHROUGHEMBED = `${id}#clickthroughEmbed`;
/** Declares the feed generator returns any types of posts. */
export const CONTENTMODEUNSPECIFIED = `${id}#contentModeUnspecified`;
/** Declares the feed generator returns posts containing app.bsky.embed.video embeds. */
export const CONTENTMODEVIDEO = `${id}#contentModeVideo`;
/** Feed item was seen by user */
export const INTERACTIONSEEN = `${id}#interactionSeen`;
/** User liked the feed item */
export const INTERACTIONLIKE = `${id}#interactionLike`;
/** User reposted the feed item */
export const INTERACTIONREPOST = `${id}#interactionRepost`;
/** User replied to the feed item */
export const INTERACTIONREPLY = `${id}#interactionReply`;
/** User quoted the feed item */
export const INTERACTIONQUOTE = `${id}#interactionQuote`;
/** User shared the feed item */
export const INTERACTIONSHARE = `${id}#interactionShare`;
