// @atproto/api/dist/client/types/app/bsky/feed/post.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'app.bsky.feed.post';
const hashMain = 'main';
export function isMain(v) {
    return is$typed(v, id, hashMain);
}
export function validateMain(v) {
    return validate(v, id, hashMain, true);
}
export { isMain as isRecord, validateMain as validateRecord, };
const hashReplyRef = 'replyRef';
export function isReplyRef(v) {
    return is$typed(v, id, hashReplyRef);
}
export function validateReplyRef(v) {
    return validate(v, id, hashReplyRef);
}
const hashEntity = 'entity';
export function isEntity(v) {
    return is$typed(v, id, hashEntity);
}
export function validateEntity(v) {
    return validate(v, id, hashEntity);
}
const hashTextSlice = 'textSlice';
export function isTextSlice(v) {
    return is$typed(v, id, hashTextSlice);
}
export function validateTextSlice(v) {
    return validate(v, id, hashTextSlice);
}
