// @atproto/api/dist/client/types/app/bsky/feed/threadgate.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'app.bsky.feed.threadgate';
const hashMain = 'main';
export function isMain(v) {
    return is$typed(v, id, hashMain);
}
export function validateMain(v) {
    return validate(v, id, hashMain, true);
}
export { isMain as isRecord, validateMain as validateRecord, };
const hashMentionRule = 'mentionRule';
export function isMentionRule(v) {
    return is$typed(v, id, hashMentionRule);
}
export function validateMentionRule(v) {
    return validate(v, id, hashMentionRule);
}
const hashFollowerRule = 'followerRule';
export function isFollowerRule(v) {
    return is$typed(v, id, hashFollowerRule);
}
export function validateFollowerRule(v) {
    return validate(v, id, hashFollowerRule);
}
const hashFollowingRule = 'followingRule';
export function isFollowingRule(v) {
    return is$typed(v, id, hashFollowingRule);
}
export function validateFollowingRule(v) {
    return validate(v, id, hashFollowingRule);
}
const hashListRule = 'listRule';
export function isListRule(v) {
    return is$typed(v, id, hashListRule);
}
export function validateListRule(v) {
    return validate(v, id, hashListRule);
}
