// @atproto/api/dist/client/types/com/atproto/admin/defs.js
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'com.atproto.admin.defs';
const hashStatusAttr = 'statusAttr';
export function isStatusAttr(v) {
    return is$typed(v, id, hashStatusAttr);
}
export function validateStatusAttr(v) {
    return validate(v, id, hashStatusAttr);
}
const hashAccountView = 'accountView';
export function isAccountView(v) {
    return is$typed(v, id, hashAccountView);
}
export function validateAccountView(v) {
    return validate(v, id, hashAccountView);
}
const hashRepoRef = 'repoRef';
export function isRepoRef(v) {
    return is$typed(v, id, hashRepoRef);
}
export function validateRepoRef(v) {
    return validate(v, id, hashRepoRef);
}
const hashRepoBlobRef = 'repoBlobRef';
export function isRepoBlobRef(v) {
    return is$typed(v, id, hashRepoBlobRef);
}
export function validateRepoBlobRef(v) {
    return validate(v, id, hashRepoBlobRef);
}
const hashThreatSignature = 'threatSignature';
export function isThreatSignature(v) {
    return is$typed(v, id, hashThreatSignature);
}
export function validateThreatSignature(v) {
    return validate(v, id, hashThreatSignature);
}
