// @atproto/api/dist/client/index.js
/**
 * GENERATED CODE - DO NOT MODIFY
 */
import { XrpcClient, } from '@atproto/xrpc';
import { schemas } from './lexicons.js';
import * as AppBskyAgeassuranceBegin from './types/app/bsky/ageassurance/begin.js';
import * as AppBskyBookmarkCreateBookmark from './types/app/bsky/bookmark/createBookmark.js';
import * as AppBskyBookmarkDeleteBookmark from './types/app/bsky/bookmark/deleteBookmark.js';
import * as AppBskyContactDismissMatch from './types/app/bsky/contact/dismissMatch.js';
import * as AppBskyContactGetMatches from './types/app/bsky/contact/getMatches.js';
import * as AppBskyContactGetSyncStatus from './types/app/bsky/contact/getSyncStatus.js';
import * as AppBskyContactImportContacts from './types/app/bsky/contact/importContacts.js';
import * as AppBskyContactRemoveData from './types/app/bsky/contact/removeData.js';
import * as AppBskyContactStartPhoneVerification from './types/app/bsky/contact/startPhoneVerification.js';
import * as AppBskyContactVerifyPhone from './types/app/bsky/contact/verifyPhone.js';
import * as AppBskyDraftCreateDraft from './types/app/bsky/draft/createDraft.js';
import * as AppBskyFeedGetActorLikes from './types/app/bsky/feed/getActorLikes.js';
import * as AppBskyFeedGetAuthorFeed from './types/app/bsky/feed/getAuthorFeed.js';
import * as AppBskyFeedGetFeed from './types/app/bsky/feed/getFeed.js';
import * as AppBskyFeedGetFeedSkeleton from './types/app/bsky/feed/getFeedSkeleton.js';
import * as AppBskyFeedGetListFeed from './types/app/bsky/feed/getListFeed.js';
import * as AppBskyFeedGetPostThread from './types/app/bsky/feed/getPostThread.js';
import * as AppBskyFeedSearchPosts from './types/app/bsky/feed/searchPosts.js';
import * as AppBskyFeedSearchPostsV2 from './types/app/bsky/feed/searchPostsV2.js';
import * as AppBskyGraphGetRelationships from './types/app/bsky/graph/getRelationships.js';
import * as AppBskyUnspeccedInitAgeAssurance from './types/app/bsky/unspecced/initAgeAssurance.js';
import * as AppBskyUnspeccedSearchActorsSkeleton from './types/app/bsky/unspecced/searchActorsSkeleton.js';
import * as AppBskyUnspeccedSearchPostsSkeleton from './types/app/bsky/unspecced/searchPostsSkeleton.js';
import * as AppBskyUnspeccedSearchStarterPacksSkeleton from './types/app/bsky/unspecced/searchStarterPacksSkeleton.js';
import * as ChatBskyConvoAcceptConvo from './types/chat/bsky/convo/acceptConvo.js';
import * as ChatBskyConvoAddReaction from './types/chat/bsky/convo/addReaction.js';
import * as ChatBskyConvoDeleteMessageForSelf from './types/chat/bsky/convo/deleteMessageForSelf.js';
import * as ChatBskyConvoGetConvo from './types/chat/bsky/convo/getConvo.js';
import * as ChatBskyConvoGetConvoForMembers from './types/chat/bsky/convo/getConvoForMembers.js';
import * as ChatBskyConvoGetConvoMembers from './types/chat/bsky/convo/getConvoMembers.js';
import * as ChatBskyConvoGetMessages from './types/chat/bsky/convo/getMessages.js';
import * as ChatBskyConvoLeaveConvo from './types/chat/bsky/convo/leaveConvo.js';
import * as ChatBskyConvoLockConvo from './types/chat/bsky/convo/lockConvo.js';
import * as ChatBskyConvoMuteConvo from './types/chat/bsky/convo/muteConvo.js';
import * as ChatBskyConvoRemoveReaction from './types/chat/bsky/convo/removeReaction.js';
import * as ChatBskyConvoSendMessage from './types/chat/bsky/convo/sendMessage.js';
import * as ChatBskyConvoSendMessageBatch from './types/chat/bsky/convo/sendMessageBatch.js';
import * as ChatBskyConvoUnlockConvo from './types/chat/bsky/convo/unlockConvo.js';
import * as ChatBskyConvoUnmuteConvo from './types/chat/bsky/convo/unmuteConvo.js';
import * as ChatBskyConvoUpdateRead from './types/chat/bsky/convo/updateRead.js';
import * as ChatBskyGroupAddMembers from './types/chat/bsky/group/addMembers.js';
import * as ChatBskyGroupApproveJoinRequest from './types/chat/bsky/group/approveJoinRequest.js';
import * as ChatBskyGroupCreateGroup from './types/chat/bsky/group/createGroup.js';
import * as ChatBskyGroupCreateJoinLink from './types/chat/bsky/group/createJoinLink.js';
import * as ChatBskyGroupDisableJoinLink from './types/chat/bsky/group/disableJoinLink.js';
import * as ChatBskyGroupEditGroup from './types/chat/bsky/group/editGroup.js';
import * as ChatBskyGroupEditJoinLink from './types/chat/bsky/group/editJoinLink.js';
import * as ChatBskyGroupEnableJoinLink from './types/chat/bsky/group/enableJoinLink.js';
import * as ChatBskyGroupListJoinRequests from './types/chat/bsky/group/listJoinRequests.js';
import * as ChatBskyGroupRejectJoinRequest from './types/chat/bsky/group/rejectJoinRequest.js';
import * as ChatBskyGroupRemoveMembers from './types/chat/bsky/group/removeMembers.js';
import * as ChatBskyGroupRequestJoin from './types/chat/bsky/group/requestJoin.js';
import * as ChatBskyGroupUpdateJoinRequestsRead from './types/chat/bsky/group/updateJoinRequestsRead.js';
import * as ChatBskyGroupWithdrawJoinRequest from './types/chat/bsky/group/withdrawJoinRequest.js';
import * as ChatBskyModerationGetConvo from './types/chat/bsky/moderation/getConvo.js';
import * as ChatBskyModerationGetConvoMembers from './types/chat/bsky/moderation/getConvoMembers.js';
import * as ComAtprotoIdentityRefreshIdentity from './types/com/atproto/identity/refreshIdentity.js';
import * as ComAtprotoIdentityResolveDid from './types/com/atproto/identity/resolveDid.js';
import * as ComAtprotoIdentityResolveHandle from './types/com/atproto/identity/resolveHandle.js';
import * as ComAtprotoIdentityResolveIdentity from './types/com/atproto/identity/resolveIdentity.js';
import * as ComAtprotoLexiconResolveLexicon from './types/com/atproto/lexicon/resolveLexicon.js';
import * as ComAtprotoRepoApplyWrites from './types/com/atproto/repo/applyWrites.js';
import * as ComAtprotoRepoCreateRecord from './types/com/atproto/repo/createRecord.js';
import * as ComAtprotoRepoDeleteRecord from './types/com/atproto/repo/deleteRecord.js';
import * as ComAtprotoRepoGetRecord from './types/com/atproto/repo/getRecord.js';
import * as ComAtprotoRepoPutRecord from './types/com/atproto/repo/putRecord.js';
import * as ComAtprotoServerConfirmEmail from './types/com/atproto/server/confirmEmail.js';
import * as ComAtprotoServerCreateAccount from './types/com/atproto/server/createAccount.js';
import * as ComAtprotoServerCreateAppPassword from './types/com/atproto/server/createAppPassword.js';
import * as ComAtprotoServerCreateSession from './types/com/atproto/server/createSession.js';
import * as ComAtprotoServerDeleteAccount from './types/com/atproto/server/deleteAccount.js';
import * as ComAtprotoServerDeleteSession from './types/com/atproto/server/deleteSession.js';
import * as ComAtprotoServerGetAccountInviteCodes from './types/com/atproto/server/getAccountInviteCodes.js';
import * as ComAtprotoServerGetServiceAuth from './types/com/atproto/server/getServiceAuth.js';
import * as ComAtprotoServerListAppPasswords from './types/com/atproto/server/listAppPasswords.js';
import * as ComAtprotoServerRefreshSession from './types/com/atproto/server/refreshSession.js';
import * as ComAtprotoServerResetPassword from './types/com/atproto/server/resetPassword.js';
import * as ComAtprotoServerUpdateEmail from './types/com/atproto/server/updateEmail.js';
import * as ComAtprotoSyncGetBlob from './types/com/atproto/sync/getBlob.js';
import * as ComAtprotoSyncGetBlocks from './types/com/atproto/sync/getBlocks.js';
import * as ComAtprotoSyncGetHead from './types/com/atproto/sync/getHead.js';
import * as ComAtprotoSyncGetHostStatus from './types/com/atproto/sync/getHostStatus.js';
import * as ComAtprotoSyncGetLatestCommit from './types/com/atproto/sync/getLatestCommit.js';
import * as ComAtprotoSyncGetRecord from './types/com/atproto/sync/getRecord.js';
import * as ComAtprotoSyncGetRepo from './types/com/atproto/sync/getRepo.js';
import * as ComAtprotoSyncGetRepoStatus from './types/com/atproto/sync/getRepoStatus.js';
import * as ComAtprotoSyncListBlobs from './types/com/atproto/sync/listBlobs.js';
import * as ComAtprotoSyncRequestCrawl from './types/com/atproto/sync/requestCrawl.js';
import * as ComAtprotoTempCheckHandleAvailability from './types/com/atproto/temp/checkHandleAvailability.js';
import * as ComAtprotoTempDereferenceScope from './types/com/atproto/temp/dereferenceScope.js';
import * as ToolsOzoneCommunicationCreateTemplate from './types/tools/ozone/communication/createTemplate.js';
import * as ToolsOzoneCommunicationUpdateTemplate from './types/tools/ozone/communication/updateTemplate.js';
import * as ToolsOzoneModerationEmitEvent from './types/tools/ozone/moderation/emitEvent.js';
import * as ToolsOzoneModerationGetAccountTimeline from './types/tools/ozone/moderation/getAccountTimeline.js';
import * as ToolsOzoneModerationGetRecord from './types/tools/ozone/moderation/getRecord.js';
import * as ToolsOzoneModerationGetRepo from './types/tools/ozone/moderation/getRepo.js';
import * as ToolsOzoneQueueAssignModerator from './types/tools/ozone/queue/assignModerator.js';
import * as ToolsOzoneQueueCreateQueue from './types/tools/ozone/queue/createQueue.js';
import * as ToolsOzoneQueueRouteReports from './types/tools/ozone/queue/routeReports.js';
import * as ToolsOzoneQueueUnassignModerator from './types/tools/ozone/queue/unassignModerator.js';
import * as ToolsOzoneReportAssignModerator from './types/tools/ozone/report/assignModerator.js';
import * as ToolsOzoneReportCreateActivity from './types/tools/ozone/report/createActivity.js';
import * as ToolsOzoneReportGetLatestReport from './types/tools/ozone/report/getLatestReport.js';
import * as ToolsOzoneReportGetReport from './types/tools/ozone/report/getReport.js';
import * as ToolsOzoneReportReassignQueue from './types/tools/ozone/report/reassignQueue.js';
import * as ToolsOzoneReportUnassignModerator from './types/tools/ozone/report/unassignModerator.js';
import * as ToolsOzoneSafelinkAddRule from './types/tools/ozone/safelink/addRule.js';
import * as ToolsOzoneSafelinkRemoveRule from './types/tools/ozone/safelink/removeRule.js';
import * as ToolsOzoneSafelinkUpdateRule from './types/tools/ozone/safelink/updateRule.js';
import * as ToolsOzoneSetDeleteSet from './types/tools/ozone/set/deleteSet.js';
import * as ToolsOzoneSetDeleteValues from './types/tools/ozone/set/deleteValues.js';
import * as ToolsOzoneSetGetValues from './types/tools/ozone/set/getValues.js';
import * as ToolsOzoneTeamAddMember from './types/tools/ozone/team/addMember.js';
import * as ToolsOzoneTeamDeleteMember from './types/tools/ozone/team/deleteMember.js';
import * as ToolsOzoneTeamUpdateMember from './types/tools/ozone/team/updateMember.js';
export * as AppBskyActorDefs from './types/app/bsky/actor/defs.js';
export * as AppBskyActorGetPreferences from './types/app/bsky/actor/getPreferences.js';
export * as AppBskyActorGetProfile from './types/app/bsky/actor/getProfile.js';
export * as AppBskyActorGetProfiles from './types/app/bsky/actor/getProfiles.js';
export * as AppBskyActorGetSuggestions from './types/app/bsky/actor/getSuggestions.js';
export * as AppBskyActorProfile from './types/app/bsky/actor/profile.js';
export * as AppBskyActorPutPreferences from './types/app/bsky/actor/putPreferences.js';
export * as AppBskyActorSearchActors from './types/app/bsky/actor/searchActors.js';
export * as AppBskyActorSearchActorsTypeahead from './types/app/bsky/actor/searchActorsTypeahead.js';
export * as AppBskyActorStatus from './types/app/bsky/actor/status.js';
export * as AppBskyAgeassuranceBegin from './types/app/bsky/ageassurance/begin.js';
export * as AppBskyAgeassuranceDefs from './types/app/bsky/ageassurance/defs.js';
export * as AppBskyAgeassuranceGetConfig from './types/app/bsky/ageassurance/getConfig.js';
export * as AppBskyAgeassuranceGetState from './types/app/bsky/ageassurance/getState.js';
export * as AppBskyBookmarkCreateBookmark from './types/app/bsky/bookmark/createBookmark.js';
export * as AppBskyBookmarkDefs from './types/app/bsky/bookmark/defs.js';
export * as AppBskyBookmarkDeleteBookmark from './types/app/bsky/bookmark/deleteBookmark.js';
export * as AppBskyBookmarkGetBookmarks from './types/app/bsky/bookmark/getBookmarks.js';
export * as AppBskyContactDefs from './types/app/bsky/contact/defs.js';
export * as AppBskyContactDismissMatch from './types/app/bsky/contact/dismissMatch.js';
export * as AppBskyContactGetMatches from './types/app/bsky/contact/getMatches.js';
export * as AppBskyContactGetSyncStatus from './types/app/bsky/contact/getSyncStatus.js';
export * as AppBskyContactImportContacts from './types/app/bsky/contact/importContacts.js';
export * as AppBskyContactRemoveData from './types/app/bsky/contact/removeData.js';
export * as AppBskyContactSendNotification from './types/app/bsky/contact/sendNotification.js';
export * as AppBskyContactStartPhoneVerification from './types/app/bsky/contact/startPhoneVerification.js';
export * as AppBskyContactVerifyPhone from './types/app/bsky/contact/verifyPhone.js';
export * as AppBskyDraftCreateDraft from './types/app/bsky/draft/createDraft.js';
export * as AppBskyDraftDefs from './types/app/bsky/draft/defs.js';
export * as AppBskyDraftDeleteDraft from './types/app/bsky/draft/deleteDraft.js';
export * as AppBskyDraftGetDrafts from './types/app/bsky/draft/getDrafts.js';
export * as AppBskyDraftUpdateDraft from './types/app/bsky/draft/updateDraft.js';
export * as AppBskyEmbedDefs from './types/app/bsky/embed/defs.js';
export * as AppBskyEmbedExternal from './types/app/bsky/embed/external.js';
export * as AppBskyEmbedGallery from './types/app/bsky/embed/gallery.js';
export * as AppBskyEmbedGetEmbedExternalView from './types/app/bsky/embed/getEmbedExternalView.js';
export * as AppBskyEmbedImages from './types/app/bsky/embed/images.js';
export * as AppBskyEmbedRecord from './types/app/bsky/embed/record.js';
export * as AppBskyEmbedRecordWithMedia from './types/app/bsky/embed/recordWithMedia.js';
export * as AppBskyEmbedVideo from './types/app/bsky/embed/video.js';
export * as AppBskyFeedDefs from './types/app/bsky/feed/defs.js';
export * as AppBskyFeedDescribeFeedGenerator from './types/app/bsky/feed/describeFeedGenerator.js';
export * as AppBskyFeedGenerator from './types/app/bsky/feed/generator.js';
export * as AppBskyFeedGetActorFeeds from './types/app/bsky/feed/getActorFeeds.js';
export * as AppBskyFeedGetActorLikes from './types/app/bsky/feed/getActorLikes.js';
export * as AppBskyFeedGetAuthorFeed from './types/app/bsky/feed/getAuthorFeed.js';
export * as AppBskyFeedGetFeed from './types/app/bsky/feed/getFeed.js';
export * as AppBskyFeedGetFeedGenerator from './types/app/bsky/feed/getFeedGenerator.js';
export * as AppBskyFeedGetFeedGenerators from './types/app/bsky/feed/getFeedGenerators.js';
export * as AppBskyFeedGetFeedSkeleton from './types/app/bsky/feed/getFeedSkeleton.js';
export * as AppBskyFeedGetLikes from './types/app/bsky/feed/getLikes.js';
export * as AppBskyFeedGetListFeed from './types/app/bsky/feed/getListFeed.js';
export * as AppBskyFeedGetPostThread from './types/app/bsky/feed/getPostThread.js';
export * as AppBskyFeedGetPosts from './types/app/bsky/feed/getPosts.js';
export * as AppBskyFeedGetQuotes from './types/app/bsky/feed/getQuotes.js';
export * as AppBskyFeedGetRepostedBy from './types/app/bsky/feed/getRepostedBy.js';
export * as AppBskyFeedGetSuggestedFeeds from './types/app/bsky/feed/getSuggestedFeeds.js';
export * as AppBskyFeedGetTimeline from './types/app/bsky/feed/getTimeline.js';
export * as AppBskyFeedLike from './types/app/bsky/feed/like.js';
export * as AppBskyFeedPost from './types/app/bsky/feed/post.js';
export * as AppBskyFeedPostgate from './types/app/bsky/feed/postgate.js';
export * as AppBskyFeedRepost from './types/app/bsky/feed/repost.js';
export * as AppBskyFeedSearchPosts from './types/app/bsky/feed/searchPosts.js';
export * as AppBskyFeedSearchPostsV2 from './types/app/bsky/feed/searchPostsV2.js';
export * as AppBskyFeedSendInteractions from './types/app/bsky/feed/sendInteractions.js';
export * as AppBskyFeedThreadgate from './types/app/bsky/feed/threadgate.js';
export * as AppBskyGraphBlock from './types/app/bsky/graph/block.js';
export * as AppBskyGraphDefs from './types/app/bsky/graph/defs.js';
export * as AppBskyGraphFollow from './types/app/bsky/graph/follow.js';
export * as AppBskyGraphGetActorStarterPacks from './types/app/bsky/graph/getActorStarterPacks.js';
export * as AppBskyGraphGetBlocks from './types/app/bsky/graph/getBlocks.js';
export * as AppBskyGraphGetFollowers from './types/app/bsky/graph/getFollowers.js';
export * as AppBskyGraphGetFollows from './types/app/bsky/graph/getFollows.js';
export * as AppBskyGraphGetKnownFollowers from './types/app/bsky/graph/getKnownFollowers.js';
export * as AppBskyGraphGetList from './types/app/bsky/graph/getList.js';
export * as AppBskyGraphGetListBlocks from './types/app/bsky/graph/getListBlocks.js';
export * as AppBskyGraphGetListMutes from './types/app/bsky/graph/getListMutes.js';
export * as AppBskyGraphGetLists from './types/app/bsky/graph/getLists.js';
export * as AppBskyGraphGetListsWithMembership from './types/app/bsky/graph/getListsWithMembership.js';
export * as AppBskyGraphGetMutes from './types/app/bsky/graph/getMutes.js';
export * as AppBskyGraphGetRelationships from './types/app/bsky/graph/getRelationships.js';
export * as AppBskyGraphGetStarterPack from './types/app/bsky/graph/getStarterPack.js';
export * as AppBskyGraphGetStarterPacks from './types/app/bsky/graph/getStarterPacks.js';
export * as AppBskyGraphGetStarterPacksWithMembership from './types/app/bsky/graph/getStarterPacksWithMembership.js';
export * as AppBskyGraphGetSuggestedFollowsByActor from './types/app/bsky/graph/getSuggestedFollowsByActor.js';
export * as AppBskyGraphList from './types/app/bsky/graph/list.js';
export * as AppBskyGraphListblock from './types/app/bsky/graph/listblock.js';
export * as AppBskyGraphListitem from './types/app/bsky/graph/listitem.js';
export * as AppBskyGraphMuteActor from './types/app/bsky/graph/muteActor.js';
export * as AppBskyGraphMuteActorList from './types/app/bsky/graph/muteActorList.js';
export * as AppBskyGraphMuteThread from './types/app/bsky/graph/muteThread.js';
export * as AppBskyGraphSearchStarterPacks from './types/app/bsky/graph/searchStarterPacks.js';
export * as AppBskyGraphStarterpack from './types/app/bsky/graph/starterpack.js';
export * as AppBskyGraphUnmuteActor from './types/app/bsky/graph/unmuteActor.js';
export * as AppBskyGraphUnmuteActorList from './types/app/bsky/graph/unmuteActorList.js';
export * as AppBskyGraphUnmuteThread from './types/app/bsky/graph/unmuteThread.js';
export * as AppBskyGraphVerification from './types/app/bsky/graph/verification.js';
export * as AppBskyLabelerDefs from './types/app/bsky/labeler/defs.js';
export * as AppBskyLabelerGetServices from './types/app/bsky/labeler/getServices.js';
export * as AppBskyLabelerService from './types/app/bsky/labeler/service.js';
export * as AppBskyNotificationDeclaration from './types/app/bsky/notification/declaration.js';
export * as AppBskyNotificationDefs from './types/app/bsky/notification/defs.js';
export * as AppBskyNotificationGetPreferences from './types/app/bsky/notification/getPreferences.js';
export * as AppBskyNotificationGetUnreadCount from './types/app/bsky/notification/getUnreadCount.js';
export * as AppBskyNotificationListActivitySubscriptions from './types/app/bsky/notification/listActivitySubscriptions.js';
export * as AppBskyNotificationListNotifications from './types/app/bsky/notification/listNotifications.js';
export * as AppBskyNotificationPutActivitySubscription from './types/app/bsky/notification/putActivitySubscription.js';
export * as AppBskyNotificationPutPreferences from './types/app/bsky/notification/putPreferences.js';
export * as AppBskyNotificationPutPreferencesV2 from './types/app/bsky/notification/putPreferencesV2.js';
export * as AppBskyNotificationRegisterPush from './types/app/bsky/notification/registerPush.js';
export * as AppBskyNotificationUnregisterPush from './types/app/bsky/notification/unregisterPush.js';
export * as AppBskyNotificationUpdateSeen from './types/app/bsky/notification/updateSeen.js';
export * as AppBskyRichtextFacet from './types/app/bsky/richtext/facet.js';
export * as AppBskyUnspeccedDefs from './types/app/bsky/unspecced/defs.js';
export * as AppBskyUnspeccedGetAgeAssuranceState from './types/app/bsky/unspecced/getAgeAssuranceState.js';
export * as AppBskyUnspeccedGetConfig from './types/app/bsky/unspecced/getConfig.js';
export * as AppBskyUnspeccedGetOnboardingSuggestedStarterPacks from './types/app/bsky/unspecced/getOnboardingSuggestedStarterPacks.js';
export * as AppBskyUnspeccedGetOnboardingSuggestedStarterPacksSkeleton from './types/app/bsky/unspecced/getOnboardingSuggestedStarterPacksSkeleton.js';
export * as AppBskyUnspeccedGetOnboardingSuggestedUsersSkeleton from './types/app/bsky/unspecced/getOnboardingSuggestedUsersSkeleton.js';
export * as AppBskyUnspeccedGetPopularFeedGenerators from './types/app/bsky/unspecced/getPopularFeedGenerators.js';
export * as AppBskyUnspeccedGetPostThreadOtherV2 from './types/app/bsky/unspecced/getPostThreadOtherV2.js';
export * as AppBskyUnspeccedGetPostThreadV2 from './types/app/bsky/unspecced/getPostThreadV2.js';
export * as AppBskyUnspeccedGetSuggestedFeeds from './types/app/bsky/unspecced/getSuggestedFeeds.js';
export * as AppBskyUnspeccedGetSuggestedFeedsSkeleton from './types/app/bsky/unspecced/getSuggestedFeedsSkeleton.js';
export * as AppBskyUnspeccedGetSuggestedOnboardingUsers from './types/app/bsky/unspecced/getSuggestedOnboardingUsers.js';
export * as AppBskyUnspeccedGetSuggestedStarterPacks from './types/app/bsky/unspecced/getSuggestedStarterPacks.js';
export * as AppBskyUnspeccedGetSuggestedStarterPacksSkeleton from './types/app/bsky/unspecced/getSuggestedStarterPacksSkeleton.js';
export * as AppBskyUnspeccedGetSuggestedUsers from './types/app/bsky/unspecced/getSuggestedUsers.js';
export * as AppBskyUnspeccedGetSuggestedUsersForDiscover from './types/app/bsky/unspecced/getSuggestedUsersForDiscover.js';
export * as AppBskyUnspeccedGetSuggestedUsersForDiscoverSkeleton from './types/app/bsky/unspecced/getSuggestedUsersForDiscoverSkeleton.js';
export * as AppBskyUnspeccedGetSuggestedUsersForExplore from './types/app/bsky/unspecced/getSuggestedUsersForExplore.js';
export * as AppBskyUnspeccedGetSuggestedUsersForExploreSkeleton from './types/app/bsky/unspecced/getSuggestedUsersForExploreSkeleton.js';
export * as AppBskyUnspeccedGetSuggestedUsersForSeeMore from './types/app/bsky/unspecced/getSuggestedUsersForSeeMore.js';
export * as AppBskyUnspeccedGetSuggestedUsersForSeeMoreSkeleton from './types/app/bsky/unspecced/getSuggestedUsersForSeeMoreSkeleton.js';
export * as AppBskyUnspeccedGetSuggestedUsersSkeleton from './types/app/bsky/unspecced/getSuggestedUsersSkeleton.js';
export * as AppBskyUnspeccedGetSuggestionsSkeleton from './types/app/bsky/unspecced/getSuggestionsSkeleton.js';
export * as AppBskyUnspeccedGetTaggedSuggestions from './types/app/bsky/unspecced/getTaggedSuggestions.js';
export * as AppBskyUnspeccedGetTrendingTopics from './types/app/bsky/unspecced/getTrendingTopics.js';
export * as AppBskyUnspeccedGetTrends from './types/app/bsky/unspecced/getTrends.js';
export * as AppBskyUnspeccedGetTrendsSkeleton from './types/app/bsky/unspecced/getTrendsSkeleton.js';
export * as AppBskyUnspeccedInitAgeAssurance from './types/app/bsky/unspecced/initAgeAssurance.js';
export * as AppBskyUnspeccedSearchActorsSkeleton from './types/app/bsky/unspecced/searchActorsSkeleton.js';
export * as AppBskyUnspeccedSearchPostsSkeleton from './types/app/bsky/unspecced/searchPostsSkeleton.js';
export * as AppBskyUnspeccedSearchStarterPacksSkeleton from './types/app/bsky/unspecced/searchStarterPacksSkeleton.js';
export * as AppBskyVideoDefs from './types/app/bsky/video/defs.js';
export * as AppBskyVideoGetJobStatus from './types/app/bsky/video/getJobStatus.js';
export * as AppBskyVideoGetUploadLimits from './types/app/bsky/video/getUploadLimits.js';
export * as AppBskyVideoUploadVideo from './types/app/bsky/video/uploadVideo.js';
export * as ChatBskyActorDeclaration from './types/chat/bsky/actor/declaration.js';
export * as ChatBskyActorDefs from './types/chat/bsky/actor/defs.js';
export * as ChatBskyActorDeleteAccount from './types/chat/bsky/actor/deleteAccount.js';
export * as ChatBskyActorExportAccountData from './types/chat/bsky/actor/exportAccountData.js';
export * as ChatBskyActorGetStatus from './types/chat/bsky/actor/getStatus.js';
export * as ChatBskyConvoAcceptConvo from './types/chat/bsky/convo/acceptConvo.js';
export * as ChatBskyConvoAddReaction from './types/chat/bsky/convo/addReaction.js';
export * as ChatBskyConvoDefs from './types/chat/bsky/convo/defs.js';
export * as ChatBskyConvoDeleteMessageForSelf from './types/chat/bsky/convo/deleteMessageForSelf.js';
export * as ChatBskyConvoGetConvo from './types/chat/bsky/convo/getConvo.js';
export * as ChatBskyConvoGetConvoAvailability from './types/chat/bsky/convo/getConvoAvailability.js';
export * as ChatBskyConvoGetConvoForMembers from './types/chat/bsky/convo/getConvoForMembers.js';
export * as ChatBskyConvoGetConvoMembers from './types/chat/bsky/convo/getConvoMembers.js';
export * as ChatBskyConvoGetLog from './types/chat/bsky/convo/getLog.js';
export * as ChatBskyConvoGetMessages from './types/chat/bsky/convo/getMessages.js';
export * as ChatBskyConvoGetUnreadCounts from './types/chat/bsky/convo/getUnreadCounts.js';
export * as ChatBskyConvoLeaveConvo from './types/chat/bsky/convo/leaveConvo.js';
export * as ChatBskyConvoListConvoRequests from './types/chat/bsky/convo/listConvoRequests.js';
export * as ChatBskyConvoListConvos from './types/chat/bsky/convo/listConvos.js';
export * as ChatBskyConvoLockConvo from './types/chat/bsky/convo/lockConvo.js';
export * as ChatBskyConvoMuteConvo from './types/chat/bsky/convo/muteConvo.js';
export * as ChatBskyConvoRemoveReaction from './types/chat/bsky/convo/removeReaction.js';
export * as ChatBskyConvoSendMessage from './types/chat/bsky/convo/sendMessage.js';
export * as ChatBskyConvoSendMessageBatch from './types/chat/bsky/convo/sendMessageBatch.js';
export * as ChatBskyConvoUnlockConvo from './types/chat/bsky/convo/unlockConvo.js';
export * as ChatBskyConvoUnmuteConvo from './types/chat/bsky/convo/unmuteConvo.js';
export * as ChatBskyConvoUpdateAllRead from './types/chat/bsky/convo/updateAllRead.js';
export * as ChatBskyConvoUpdateRead from './types/chat/bsky/convo/updateRead.js';
export * as ChatBskyEmbedJoinLink from './types/chat/bsky/embed/joinLink.js';
export * as ChatBskyGroupAddMembers from './types/chat/bsky/group/addMembers.js';
export * as ChatBskyGroupApproveJoinRequest from './types/chat/bsky/group/approveJoinRequest.js';
export * as ChatBskyGroupCreateGroup from './types/chat/bsky/group/createGroup.js';
export * as ChatBskyGroupCreateJoinLink from './types/chat/bsky/group/createJoinLink.js';
export * as ChatBskyGroupDefs from './types/chat/bsky/group/defs.js';
export * as ChatBskyGroupDisableJoinLink from './types/chat/bsky/group/disableJoinLink.js';
export * as ChatBskyGroupEditGroup from './types/chat/bsky/group/editGroup.js';
export * as ChatBskyGroupEditJoinLink from './types/chat/bsky/group/editJoinLink.js';
export * as ChatBskyGroupEnableJoinLink from './types/chat/bsky/group/enableJoinLink.js';
export * as ChatBskyGroupGetJoinLinkPreviews from './types/chat/bsky/group/getJoinLinkPreviews.js';
export * as ChatBskyGroupListJoinRequests from './types/chat/bsky/group/listJoinRequests.js';
export * as ChatBskyGroupListMutualGroups from './types/chat/bsky/group/listMutualGroups.js';
export * as ChatBskyGroupRejectJoinRequest from './types/chat/bsky/group/rejectJoinRequest.js';
export * as ChatBskyGroupRemoveMembers from './types/chat/bsky/group/removeMembers.js';
export * as ChatBskyGroupRequestJoin from './types/chat/bsky/group/requestJoin.js';
export * as ChatBskyGroupUpdateJoinRequestsRead from './types/chat/bsky/group/updateJoinRequestsRead.js';
export * as ChatBskyGroupWithdrawJoinRequest from './types/chat/bsky/group/withdrawJoinRequest.js';
export * as ChatBskyModerationDefs from './types/chat/bsky/moderation/defs.js';
export * as ChatBskyModerationGetActorMetadata from './types/chat/bsky/moderation/getActorMetadata.js';
export * as ChatBskyModerationGetConvo from './types/chat/bsky/moderation/getConvo.js';
export * as ChatBskyModerationGetConvoMembers from './types/chat/bsky/moderation/getConvoMembers.js';
export * as ChatBskyModerationGetConvos from './types/chat/bsky/moderation/getConvos.js';
export * as ChatBskyModerationGetMessageContext from './types/chat/bsky/moderation/getMessageContext.js';
export * as ChatBskyModerationSubscribeModEvents from './types/chat/bsky/moderation/subscribeModEvents.js';
export * as ChatBskyModerationUpdateActorAccess from './types/chat/bsky/moderation/updateActorAccess.js';
export * as ChatBskyNotificationDefs from './types/chat/bsky/notification/defs.js';
export * as ChatBskyNotificationGetPreferences from './types/chat/bsky/notification/getPreferences.js';
export * as ChatBskyNotificationPutPreferences from './types/chat/bsky/notification/putPreferences.js';
export * as ComAtprotoAdminDefs from './types/com/atproto/admin/defs.js';
export * as ComAtprotoAdminDeleteAccount from './types/com/atproto/admin/deleteAccount.js';
export * as ComAtprotoAdminDisableAccountInvites from './types/com/atproto/admin/disableAccountInvites.js';
export * as ComAtprotoAdminDisableInviteCodes from './types/com/atproto/admin/disableInviteCodes.js';
export * as ComAtprotoAdminEnableAccountInvites from './types/com/atproto/admin/enableAccountInvites.js';
export * as ComAtprotoAdminGetAccountInfo from './types/com/atproto/admin/getAccountInfo.js';
export * as ComAtprotoAdminGetAccountInfos from './types/com/atproto/admin/getAccountInfos.js';
export * as ComAtprotoAdminGetInviteCodes from './types/com/atproto/admin/getInviteCodes.js';
export * as ComAtprotoAdminGetSubjectStatus from './types/com/atproto/admin/getSubjectStatus.js';
export * as ComAtprotoAdminSearchAccounts from './types/com/atproto/admin/searchAccounts.js';
export * as ComAtprotoAdminSendEmail from './types/com/atproto/admin/sendEmail.js';
export * as ComAtprotoAdminUpdateAccountEmail from './types/com/atproto/admin/updateAccountEmail.js';
export * as ComAtprotoAdminUpdateAccountHandle from './types/com/atproto/admin/updateAccountHandle.js';
export * as ComAtprotoAdminUpdateAccountPassword from './types/com/atproto/admin/updateAccountPassword.js';
export * as ComAtprotoAdminUpdateAccountSigningKey from './types/com/atproto/admin/updateAccountSigningKey.js';
export * as ComAtprotoAdminUpdateSubjectStatus from './types/com/atproto/admin/updateSubjectStatus.js';
export * as ComAtprotoIdentityDefs from './types/com/atproto/identity/defs.js';
export * as ComAtprotoIdentityGetRecommendedDidCredentials from './types/com/atproto/identity/getRecommendedDidCredentials.js';
export * as ComAtprotoIdentityRefreshIdentity from './types/com/atproto/identity/refreshIdentity.js';
export * as ComAtprotoIdentityRequestPlcOperationSignature from './types/com/atproto/identity/requestPlcOperationSignature.js';
export * as ComAtprotoIdentityResolveDid from './types/com/atproto/identity/resolveDid.js';
export * as ComAtprotoIdentityResolveHandle from './types/com/atproto/identity/resolveHandle.js';
export * as ComAtprotoIdentityResolveIdentity from './types/com/atproto/identity/resolveIdentity.js';
export * as ComAtprotoIdentitySignPlcOperation from './types/com/atproto/identity/signPlcOperation.js';
export * as ComAtprotoIdentitySubmitPlcOperation from './types/com/atproto/identity/submitPlcOperation.js';
export * as ComAtprotoIdentityUpdateHandle from './types/com/atproto/identity/updateHandle.js';
export * as ComAtprotoLabelDefs from './types/com/atproto/label/defs.js';
export * as ComAtprotoLabelQueryLabels from './types/com/atproto/label/queryLabels.js';
export * as ComAtprotoLabelSubscribeLabels from './types/com/atproto/label/subscribeLabels.js';
export * as ComAtprotoLexiconResolveLexicon from './types/com/atproto/lexicon/resolveLexicon.js';
export * as ComAtprotoLexiconSchema from './types/com/atproto/lexicon/schema.js';
export * as ComAtprotoModerationCreateReport from './types/com/atproto/moderation/createReport.js';
export * as ComAtprotoModerationDefs from './types/com/atproto/moderation/defs.js';
export * as ComAtprotoRepoApplyWrites from './types/com/atproto/repo/applyWrites.js';
export * as ComAtprotoRepoCreateRecord from './types/com/atproto/repo/createRecord.js';
export * as ComAtprotoRepoDefs from './types/com/atproto/repo/defs.js';
export * as ComAtprotoRepoDeleteRecord from './types/com/atproto/repo/deleteRecord.js';
export * as ComAtprotoRepoDescribeRepo from './types/com/atproto/repo/describeRepo.js';
export * as ComAtprotoRepoGetRecord from './types/com/atproto/repo/getRecord.js';
export * as ComAtprotoRepoImportRepo from './types/com/atproto/repo/importRepo.js';
export * as ComAtprotoRepoListMissingBlobs from './types/com/atproto/repo/listMissingBlobs.js';
export * as ComAtprotoRepoListRecords from './types/com/atproto/repo/listRecords.js';
export * as ComAtprotoRepoPutRecord from './types/com/atproto/repo/putRecord.js';
export * as ComAtprotoRepoStrongRef from './types/com/atproto/repo/strongRef.js';
export * as ComAtprotoRepoUploadBlob from './types/com/atproto/repo/uploadBlob.js';
export * as ComAtprotoServerActivateAccount from './types/com/atproto/server/activateAccount.js';
export * as ComAtprotoServerCheckAccountStatus from './types/com/atproto/server/checkAccountStatus.js';
export * as ComAtprotoServerConfirmEmail from './types/com/atproto/server/confirmEmail.js';
export * as ComAtprotoServerCreateAccount from './types/com/atproto/server/createAccount.js';
export * as ComAtprotoServerCreateAppPassword from './types/com/atproto/server/createAppPassword.js';
export * as ComAtprotoServerCreateInviteCode from './types/com/atproto/server/createInviteCode.js';
export * as ComAtprotoServerCreateInviteCodes from './types/com/atproto/server/createInviteCodes.js';
export * as ComAtprotoServerCreateSession from './types/com/atproto/server/createSession.js';
export * as ComAtprotoServerDeactivateAccount from './types/com/atproto/server/deactivateAccount.js';
export * as ComAtprotoServerDefs from './types/com/atproto/server/defs.js';
export * as ComAtprotoServerDeleteAccount from './types/com/atproto/server/deleteAccount.js';
export * as ComAtprotoServerDeleteSession from './types/com/atproto/server/deleteSession.js';
export * as ComAtprotoServerDescribeServer from './types/com/atproto/server/describeServer.js';
export * as ComAtprotoServerGetAccountInviteCodes from './types/com/atproto/server/getAccountInviteCodes.js';
export * as ComAtprotoServerGetServiceAuth from './types/com/atproto/server/getServiceAuth.js';
export * as ComAtprotoServerGetSession from './types/com/atproto/server/getSession.js';
export * as ComAtprotoServerListAppPasswords from './types/com/atproto/server/listAppPasswords.js';
export * as ComAtprotoServerRefreshSession from './types/com/atproto/server/refreshSession.js';
export * as ComAtprotoServerRequestAccountDelete from './types/com/atproto/server/requestAccountDelete.js';
export * as ComAtprotoServerRequestEmailConfirmation from './types/com/atproto/server/requestEmailConfirmation.js';
export * as ComAtprotoServerRequestEmailUpdate from './types/com/atproto/server/requestEmailUpdate.js';
export * as ComAtprotoServerRequestPasswordReset from './types/com/atproto/server/requestPasswordReset.js';
export * as ComAtprotoServerReserveSigningKey from './types/com/atproto/server/reserveSigningKey.js';
export * as ComAtprotoServerResetPassword from './types/com/atproto/server/resetPassword.js';
export * as ComAtprotoServerRevokeAppPassword from './types/com/atproto/server/revokeAppPassword.js';
export * as ComAtprotoServerUpdateEmail from './types/com/atproto/server/updateEmail.js';
export * as ComAtprotoSyncDefs from './types/com/atproto/sync/defs.js';
export * as ComAtprotoSyncGetBlob from './types/com/atproto/sync/getBlob.js';
export * as ComAtprotoSyncGetBlocks from './types/com/atproto/sync/getBlocks.js';
export * as ComAtprotoSyncGetCheckout from './types/com/atproto/sync/getCheckout.js';
export * as ComAtprotoSyncGetHead from './types/com/atproto/sync/getHead.js';
export * as ComAtprotoSyncGetHostStatus from './types/com/atproto/sync/getHostStatus.js';
export * as ComAtprotoSyncGetLatestCommit from './types/com/atproto/sync/getLatestCommit.js';
export * as ComAtprotoSyncGetRecord from './types/com/atproto/sync/getRecord.js';
export * as ComAtprotoSyncGetRepo from './types/com/atproto/sync/getRepo.js';
export * as ComAtprotoSyncGetRepoStatus from './types/com/atproto/sync/getRepoStatus.js';
export * as ComAtprotoSyncListBlobs from './types/com/atproto/sync/listBlobs.js';
export * as ComAtprotoSyncListHosts from './types/com/atproto/sync/listHosts.js';
export * as ComAtprotoSyncListRepos from './types/com/atproto/sync/listRepos.js';
export * as ComAtprotoSyncListReposByCollection from './types/com/atproto/sync/listReposByCollection.js';
export * as ComAtprotoSyncNotifyOfUpdate from './types/com/atproto/sync/notifyOfUpdate.js';
export * as ComAtprotoSyncRequestCrawl from './types/com/atproto/sync/requestCrawl.js';
export * as ComAtprotoSyncSubscribeRepos from './types/com/atproto/sync/subscribeRepos.js';
export * as ComAtprotoTempAddReservedHandle from './types/com/atproto/temp/addReservedHandle.js';
export * as ComAtprotoTempCheckHandleAvailability from './types/com/atproto/temp/checkHandleAvailability.js';
export * as ComAtprotoTempCheckSignupQueue from './types/com/atproto/temp/checkSignupQueue.js';
export * as ComAtprotoTempDereferenceScope from './types/com/atproto/temp/dereferenceScope.js';
export * as ComAtprotoTempFetchLabels from './types/com/atproto/temp/fetchLabels.js';
export * as ComAtprotoTempRequestPhoneVerification from './types/com/atproto/temp/requestPhoneVerification.js';
export * as ComAtprotoTempRevokeAccountCredentials from './types/com/atproto/temp/revokeAccountCredentials.js';
export * as ComGermnetworkDeclaration from './types/com/germnetwork/declaration.js';
export * as ToolsOzoneCommunicationCreateTemplate from './types/tools/ozone/communication/createTemplate.js';
export * as ToolsOzoneCommunicationDefs from './types/tools/ozone/communication/defs.js';
export * as ToolsOzoneCommunicationDeleteTemplate from './types/tools/ozone/communication/deleteTemplate.js';
export * as ToolsOzoneCommunicationListTemplates from './types/tools/ozone/communication/listTemplates.js';
export * as ToolsOzoneCommunicationUpdateTemplate from './types/tools/ozone/communication/updateTemplate.js';
export * as ToolsOzoneHostingGetAccountHistory from './types/tools/ozone/hosting/getAccountHistory.js';
export * as ToolsOzoneModerationCancelScheduledActions from './types/tools/ozone/moderation/cancelScheduledActions.js';
export * as ToolsOzoneModerationDefs from './types/tools/ozone/moderation/defs.js';
export * as ToolsOzoneModerationEmitEvent from './types/tools/ozone/moderation/emitEvent.js';
export * as ToolsOzoneModerationGetAccountTimeline from './types/tools/ozone/moderation/getAccountTimeline.js';
export * as ToolsOzoneModerationGetEvent from './types/tools/ozone/moderation/getEvent.js';
export * as ToolsOzoneModerationGetRecord from './types/tools/ozone/moderation/getRecord.js';
export * as ToolsOzoneModerationGetRecords from './types/tools/ozone/moderation/getRecords.js';
export * as ToolsOzoneModerationGetRepo from './types/tools/ozone/moderation/getRepo.js';
export * as ToolsOzoneModerationGetReporterStats from './types/tools/ozone/moderation/getReporterStats.js';
export * as ToolsOzoneModerationGetRepos from './types/tools/ozone/moderation/getRepos.js';
export * as ToolsOzoneModerationGetSubjects from './types/tools/ozone/moderation/getSubjects.js';
export * as ToolsOzoneModerationListScheduledActions from './types/tools/ozone/moderation/listScheduledActions.js';
export * as ToolsOzoneModerationQueryEvents from './types/tools/ozone/moderation/queryEvents.js';
export * as ToolsOzoneModerationQueryStatuses from './types/tools/ozone/moderation/queryStatuses.js';
export * as ToolsOzoneModerationScheduleAction from './types/tools/ozone/moderation/scheduleAction.js';
export * as ToolsOzoneModerationSearchRepos from './types/tools/ozone/moderation/searchRepos.js';
export * as ToolsOzoneQueueAssignModerator from './types/tools/ozone/queue/assignModerator.js';
export * as ToolsOzoneQueueCreateQueue from './types/tools/ozone/queue/createQueue.js';
export * as ToolsOzoneQueueDefs from './types/tools/ozone/queue/defs.js';
export * as ToolsOzoneQueueDeleteQueue from './types/tools/ozone/queue/deleteQueue.js';
export * as ToolsOzoneQueueGetAssignments from './types/tools/ozone/queue/getAssignments.js';
export * as ToolsOzoneQueueListQueues from './types/tools/ozone/queue/listQueues.js';
export * as ToolsOzoneQueueRouteReports from './types/tools/ozone/queue/routeReports.js';
export * as ToolsOzoneQueueUnassignModerator from './types/tools/ozone/queue/unassignModerator.js';
export * as ToolsOzoneQueueUpdateQueue from './types/tools/ozone/queue/updateQueue.js';
export * as ToolsOzoneReportAssignModerator from './types/tools/ozone/report/assignModerator.js';
export * as ToolsOzoneReportCreateActivity from './types/tools/ozone/report/createActivity.js';
export * as ToolsOzoneReportDefs from './types/tools/ozone/report/defs.js';
export * as ToolsOzoneReportGetAssignments from './types/tools/ozone/report/getAssignments.js';
export * as ToolsOzoneReportGetHistoricalStats from './types/tools/ozone/report/getHistoricalStats.js';
export * as ToolsOzoneReportGetLatestReport from './types/tools/ozone/report/getLatestReport.js';
export * as ToolsOzoneReportGetLiveStats from './types/tools/ozone/report/getLiveStats.js';
export * as ToolsOzoneReportGetReport from './types/tools/ozone/report/getReport.js';
export * as ToolsOzoneReportListActivities from './types/tools/ozone/report/listActivities.js';
export * as ToolsOzoneReportQueryActivities from './types/tools/ozone/report/queryActivities.js';
export * as ToolsOzoneReportQueryReports from './types/tools/ozone/report/queryReports.js';
export * as ToolsOzoneReportReassignQueue from './types/tools/ozone/report/reassignQueue.js';
export * as ToolsOzoneReportRefreshStats from './types/tools/ozone/report/refreshStats.js';
export * as ToolsOzoneReportUnassignModerator from './types/tools/ozone/report/unassignModerator.js';
export * as ToolsOzoneSafelinkAddRule from './types/tools/ozone/safelink/addRule.js';
export * as ToolsOzoneSafelinkDefs from './types/tools/ozone/safelink/defs.js';
export * as ToolsOzoneSafelinkQueryEvents from './types/tools/ozone/safelink/queryEvents.js';
export * as ToolsOzoneSafelinkQueryRules from './types/tools/ozone/safelink/queryRules.js';
export * as ToolsOzoneSafelinkRemoveRule from './types/tools/ozone/safelink/removeRule.js';
export * as ToolsOzoneSafelinkUpdateRule from './types/tools/ozone/safelink/updateRule.js';
export * as ToolsOzoneServerGetConfig from './types/tools/ozone/server/getConfig.js';
export * as ToolsOzoneSetAddValues from './types/tools/ozone/set/addValues.js';
export * as ToolsOzoneSetDefs from './types/tools/ozone/set/defs.js';
export * as ToolsOzoneSetDeleteSet from './types/tools/ozone/set/deleteSet.js';
export * as ToolsOzoneSetDeleteValues from './types/tools/ozone/set/deleteValues.js';
export * as ToolsOzoneSetGetValues from './types/tools/ozone/set/getValues.js';
export * as ToolsOzoneSetQuerySets from './types/tools/ozone/set/querySets.js';
export * as ToolsOzoneSetUpsertSet from './types/tools/ozone/set/upsertSet.js';
export * as ToolsOzoneSettingDefs from './types/tools/ozone/setting/defs.js';
export * as ToolsOzoneSettingListOptions from './types/tools/ozone/setting/listOptions.js';
export * as ToolsOzoneSettingRemoveOptions from './types/tools/ozone/setting/removeOptions.js';
export * as ToolsOzoneSettingUpsertOption from './types/tools/ozone/setting/upsertOption.js';
export * as ToolsOzoneSignatureDefs from './types/tools/ozone/signature/defs.js';
export * as ToolsOzoneSignatureFindCorrelation from './types/tools/ozone/signature/findCorrelation.js';
export * as ToolsOzoneSignatureFindRelatedAccounts from './types/tools/ozone/signature/findRelatedAccounts.js';
export * as ToolsOzoneSignatureSearchAccounts from './types/tools/ozone/signature/searchAccounts.js';
export * as ToolsOzoneTeamAddMember from './types/tools/ozone/team/addMember.js';
export * as ToolsOzoneTeamDefs from './types/tools/ozone/team/defs.js';
export * as ToolsOzoneTeamDeleteMember from './types/tools/ozone/team/deleteMember.js';
export * as ToolsOzoneTeamListMembers from './types/tools/ozone/team/listMembers.js';
export * as ToolsOzoneTeamUpdateMember from './types/tools/ozone/team/updateMember.js';
export * as ToolsOzoneVerificationDefs from './types/tools/ozone/verification/defs.js';
export * as ToolsOzoneVerificationGrantVerifications from './types/tools/ozone/verification/grantVerifications.js';
export * as ToolsOzoneVerificationListVerifications from './types/tools/ozone/verification/listVerifications.js';
export * as ToolsOzoneVerificationRevokeVerifications from './types/tools/ozone/verification/revokeVerifications.js';
export const APP_BSKY_ACTOR = {
    StatusLive: 'app.bsky.actor.status#live',
};
export const APP_BSKY_FEED = {
    DefsRequestLess: 'app.bsky.feed.defs#requestLess',
    DefsRequestMore: 'app.bsky.feed.defs#requestMore',
    DefsClickthroughItem: 'app.bsky.feed.defs#clickthroughItem',
    DefsClickthroughAuthor: 'app.bsky.feed.defs#clickthroughAuthor',
    DefsClickthroughReposter: 'app.bsky.feed.defs#clickthroughReposter',
    DefsClickthroughEmbed: 'app.bsky.feed.defs#clickthroughEmbed',
    DefsContentModeUnspecified: 'app.bsky.feed.defs#contentModeUnspecified',
    DefsContentModeVideo: 'app.bsky.feed.defs#contentModeVideo',
    DefsInteractionSeen: 'app.bsky.feed.defs#interactionSeen',
    DefsInteractionLike: 'app.bsky.feed.defs#interactionLike',
    DefsInteractionRepost: 'app.bsky.feed.defs#interactionRepost',
    DefsInteractionReply: 'app.bsky.feed.defs#interactionReply',
    DefsInteractionQuote: 'app.bsky.feed.defs#interactionQuote',
    DefsInteractionShare: 'app.bsky.feed.defs#interactionShare',
};
export const APP_BSKY_GRAPH = {
    DefsModlist: 'app.bsky.graph.defs#modlist',
    DefsCuratelist: 'app.bsky.graph.defs#curatelist',
    DefsReferencelist: 'app.bsky.graph.defs#referencelist',
};
export const COM_ATPROTO_MODERATION = {
    DefsReasonSpam: 'com.atproto.moderation.defs#reasonSpam',
    DefsReasonViolation: 'com.atproto.moderation.defs#reasonViolation',
    DefsReasonMisleading: 'com.atproto.moderation.defs#reasonMisleading',
    DefsReasonSexual: 'com.atproto.moderation.defs#reasonSexual',
    DefsReasonRude: 'com.atproto.moderation.defs#reasonRude',
    DefsReasonOther: 'com.atproto.moderation.defs#reasonOther',
    DefsReasonAppeal: 'com.atproto.moderation.defs#reasonAppeal',
};
export const TOOLS_OZONE_MODERATION = {
    DefsReviewOpen: 'tools.ozone.moderation.defs#reviewOpen',
    DefsReviewEscalated: 'tools.ozone.moderation.defs#reviewEscalated',
    DefsReviewClosed: 'tools.ozone.moderation.defs#reviewClosed',
    DefsReviewNone: 'tools.ozone.moderation.defs#reviewNone',
    DefsTimelineEventPlcCreate: 'tools.ozone.moderation.defs#timelineEventPlcCreate',
    DefsTimelineEventPlcOperation: 'tools.ozone.moderation.defs#timelineEventPlcOperation',
    DefsTimelineEventPlcTombstone: 'tools.ozone.moderation.defs#timelineEventPlcTombstone',
};
export const TOOLS_OZONE_REPORT = {
    DefsReasonAppeal: 'tools.ozone.report.defs#reasonAppeal',
    DefsReasonOther: 'tools.ozone.report.defs#reasonOther',
    DefsReasonViolenceAnimal: 'tools.ozone.report.defs#reasonViolenceAnimal',
    DefsReasonViolenceThreats: 'tools.ozone.report.defs#reasonViolenceThreats',
    DefsReasonViolenceGraphicContent: 'tools.ozone.report.defs#reasonViolenceGraphicContent',
    DefsReasonViolenceGlorification: 'tools.ozone.report.defs#reasonViolenceGlorification',
    DefsReasonViolenceExtremistContent: 'tools.ozone.report.defs#reasonViolenceExtremistContent',
    DefsReasonViolenceTrafficking: 'tools.ozone.report.defs#reasonViolenceTrafficking',
    DefsReasonViolenceOther: 'tools.ozone.report.defs#reasonViolenceOther',
    DefsReasonSexualAbuseContent: 'tools.ozone.report.defs#reasonSexualAbuseContent',
    DefsReasonSexualNCII: 'tools.ozone.report.defs#reasonSexualNCII',
    DefsReasonSexualDeepfake: 'tools.ozone.report.defs#reasonSexualDeepfake',
    DefsReasonSexualAnimal: 'tools.ozone.report.defs#reasonSexualAnimal',
    DefsReasonSexualUnlabeled: 'tools.ozone.report.defs#reasonSexualUnlabeled',
    DefsReasonSexualOther: 'tools.ozone.report.defs#reasonSexualOther',
    DefsReasonChildSafetyCSAM: 'tools.ozone.report.defs#reasonChildSafetyCSAM',
    DefsReasonChildSafetyGroom: 'tools.ozone.report.defs#reasonChildSafetyGroom',
    DefsReasonChildSafetyPrivacy: 'tools.ozone.report.defs#reasonChildSafetyPrivacy',
    DefsReasonChildSafetyHarassment: 'tools.ozone.report.defs#reasonChildSafetyHarassment',
    DefsReasonChildSafetyOther: 'tools.ozone.report.defs#reasonChildSafetyOther',
    DefsReasonHarassmentTroll: 'tools.ozone.report.defs#reasonHarassmentTroll',
    DefsReasonHarassmentTargeted: 'tools.ozone.report.defs#reasonHarassmentTargeted',
    DefsReasonHarassmentHateSpeech: 'tools.ozone.report.defs#reasonHarassmentHateSpeech',
    DefsReasonHarassmentDoxxing: 'tools.ozone.report.defs#reasonHarassmentDoxxing',
    DefsReasonHarassmentOther: 'tools.ozone.report.defs#reasonHarassmentOther',
    DefsReasonMisleadingBot: 'tools.ozone.report.defs#reasonMisleadingBot',
    DefsReasonMisleadingImpersonation: 'tools.ozone.report.defs#reasonMisleadingImpersonation',
    DefsReasonMisleadingSpam: 'tools.ozone.report.defs#reasonMisleadingSpam',
    DefsReasonMisleadingScam: 'tools.ozone.report.defs#reasonMisleadingScam',
    DefsReasonMisleadingElections: 'tools.ozone.report.defs#reasonMisleadingElections',
    DefsReasonMisleadingOther: 'tools.ozone.report.defs#reasonMisleadingOther',
    DefsReasonRuleSiteSecurity: 'tools.ozone.report.defs#reasonRuleSiteSecurity',
    DefsReasonRuleProhibitedSales: 'tools.ozone.report.defs#reasonRuleProhibitedSales',
    DefsReasonRuleBanEvasion: 'tools.ozone.report.defs#reasonRuleBanEvasion',
    DefsReasonRuleOther: 'tools.ozone.report.defs#reasonRuleOther',
    DefsReasonSelfHarmContent: 'tools.ozone.report.defs#reasonSelfHarmContent',
    DefsReasonSelfHarmED: 'tools.ozone.report.defs#reasonSelfHarmED',
    DefsReasonSelfHarmStunts: 'tools.ozone.report.defs#reasonSelfHarmStunts',
    DefsReasonSelfHarmSubstances: 'tools.ozone.report.defs#reasonSelfHarmSubstances',
    DefsReasonSelfHarmOther: 'tools.ozone.report.defs#reasonSelfHarmOther',
};
export const TOOLS_OZONE_TEAM = {
    DefsRoleAdmin: 'tools.ozone.team.defs#roleAdmin',
    DefsRoleModerator: 'tools.ozone.team.defs#roleModerator',
    DefsRoleTriage: 'tools.ozone.team.defs#roleTriage',
    DefsRoleVerifier: 'tools.ozone.team.defs#roleVerifier',
};
export class AtpBaseClient extends XrpcClient {
    constructor(options) {
        super(options, schemas);
        this.app = new AppNS(this);
        this.chat = new ChatNS(this);
        this.com = new ComNS(this);
        this.tools = new ToolsNS(this);
    }
    /** @deprecated use `this` instead */
    get xrpc() {
        return this;
    }
}
export class AppNS {
    constructor(client) {
        this._client = client;
        this.bsky = new AppBskyNS(client);
    }
}
export class AppBskyNS {
    constructor(client) {
        this._client = client;
        this.actor = new AppBskyActorNS(client);
        this.ageassurance = new AppBskyAgeassuranceNS(client);
        this.bookmark = new AppBskyBookmarkNS(client);
        this.contact = new AppBskyContactNS(client);
        this.draft = new AppBskyDraftNS(client);
        this.embed = new AppBskyEmbedNS(client);
        this.feed = new AppBskyFeedNS(client);
        this.graph = new AppBskyGraphNS(client);
        this.labeler = new AppBskyLabelerNS(client);
        this.notification = new AppBskyNotificationNS(client);
        this.richtext = new AppBskyRichtextNS(client);
        this.unspecced = new AppBskyUnspeccedNS(client);
        this.video = new AppBskyVideoNS(client);
    }
}
export class AppBskyActorNS {
    constructor(client) {
        this._client = client;
        this.profile = new AppBskyActorProfileRecord(client);
        this.status = new AppBskyActorStatusRecord(client);
    }
    getPreferences(params, opts) {
        return this._client.call('app.bsky.actor.getPreferences', params, undefined, opts);
    }
    getProfile(params, opts) {
        return this._client.call('app.bsky.actor.getProfile', params, undefined, opts);
    }
    getProfiles(params, opts) {
        return this._client.call('app.bsky.actor.getProfiles', params, undefined, opts);
    }
    getSuggestions(params, opts) {
        return this._client.call('app.bsky.actor.getSuggestions', params, undefined, opts);
    }
    putPreferences(data, opts) {
        return this._client.call('app.bsky.actor.putPreferences', opts?.qp, data, opts);
    }
    searchActors(params, opts) {
        return this._client.call('app.bsky.actor.searchActors', params, undefined, opts);
    }
    searchActorsTypeahead(params, opts) {
        return this._client.call('app.bsky.actor.searchActorsTypeahead', params, undefined, opts);
    }
}
export class AppBskyActorProfileRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.actor.profile',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.actor.profile',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.actor.profile';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, {
            collection,
            rkey: 'self',
            ...params,
            record: { ...record, $type: collection },
        }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.actor.profile';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.actor.profile', ...params }, { headers });
    }
}
export class AppBskyActorStatusRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.actor.status',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.actor.status',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.actor.status';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, {
            collection,
            rkey: 'self',
            ...params,
            record: { ...record, $type: collection },
        }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.actor.status';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.actor.status', ...params }, { headers });
    }
}
export class AppBskyAgeassuranceNS {
    constructor(client) {
        this._client = client;
    }
    begin(data, opts) {
        return this._client
            .call('app.bsky.ageassurance.begin', opts?.qp, data, opts)
            .catch((e) => {
            throw AppBskyAgeassuranceBegin.toKnownErr(e);
        });
    }
    getConfig(params, opts) {
        return this._client.call('app.bsky.ageassurance.getConfig', params, undefined, opts);
    }
    getState(params, opts) {
        return this._client.call('app.bsky.ageassurance.getState', params, undefined, opts);
    }
}
export class AppBskyBookmarkNS {
    constructor(client) {
        this._client = client;
    }
    createBookmark(data, opts) {
        return this._client
            .call('app.bsky.bookmark.createBookmark', opts?.qp, data, opts)
            .catch((e) => {
            throw AppBskyBookmarkCreateBookmark.toKnownErr(e);
        });
    }
    deleteBookmark(data, opts) {
        return this._client
            .call('app.bsky.bookmark.deleteBookmark', opts?.qp, data, opts)
            .catch((e) => {
            throw AppBskyBookmarkDeleteBookmark.toKnownErr(e);
        });
    }
    getBookmarks(params, opts) {
        return this._client.call('app.bsky.bookmark.getBookmarks', params, undefined, opts);
    }
}
export class AppBskyContactNS {
    constructor(client) {
        this._client = client;
    }
    dismissMatch(data, opts) {
        return this._client
            .call('app.bsky.contact.dismissMatch', opts?.qp, data, opts)
            .catch((e) => {
            throw AppBskyContactDismissMatch.toKnownErr(e);
        });
    }
    getMatches(params, opts) {
        return this._client
            .call('app.bsky.contact.getMatches', params, undefined, opts)
            .catch((e) => {
            throw AppBskyContactGetMatches.toKnownErr(e);
        });
    }
    getSyncStatus(params, opts) {
        return this._client
            .call('app.bsky.contact.getSyncStatus', params, undefined, opts)
            .catch((e) => {
            throw AppBskyContactGetSyncStatus.toKnownErr(e);
        });
    }
    importContacts(data, opts) {
        return this._client
            .call('app.bsky.contact.importContacts', opts?.qp, data, opts)
            .catch((e) => {
            throw AppBskyContactImportContacts.toKnownErr(e);
        });
    }
    removeData(data, opts) {
        return this._client
            .call('app.bsky.contact.removeData', opts?.qp, data, opts)
            .catch((e) => {
            throw AppBskyContactRemoveData.toKnownErr(e);
        });
    }
    sendNotification(data, opts) {
        return this._client.call('app.bsky.contact.sendNotification', opts?.qp, data, opts);
    }
    startPhoneVerification(data, opts) {
        return this._client
            .call('app.bsky.contact.startPhoneVerification', opts?.qp, data, opts)
            .catch((e) => {
            throw AppBskyContactStartPhoneVerification.toKnownErr(e);
        });
    }
    verifyPhone(data, opts) {
        return this._client
            .call('app.bsky.contact.verifyPhone', opts?.qp, data, opts)
            .catch((e) => {
            throw AppBskyContactVerifyPhone.toKnownErr(e);
        });
    }
}
export class AppBskyDraftNS {
    constructor(client) {
        this._client = client;
    }
    createDraft(data, opts) {
        return this._client
            .call('app.bsky.draft.createDraft', opts?.qp, data, opts)
            .catch((e) => {
            throw AppBskyDraftCreateDraft.toKnownErr(e);
        });
    }
    deleteDraft(data, opts) {
        return this._client.call('app.bsky.draft.deleteDraft', opts?.qp, data, opts);
    }
    getDrafts(params, opts) {
        return this._client.call('app.bsky.draft.getDrafts', params, undefined, opts);
    }
    updateDraft(data, opts) {
        return this._client.call('app.bsky.draft.updateDraft', opts?.qp, data, opts);
    }
}
export class AppBskyEmbedNS {
    constructor(client) {
        this._client = client;
    }
    getEmbedExternalView(params, opts) {
        return this._client.call('app.bsky.embed.getEmbedExternalView', params, undefined, opts);
    }
}
export class AppBskyFeedNS {
    constructor(client) {
        this._client = client;
        this.generator = new AppBskyFeedGeneratorRecord(client);
        this.like = new AppBskyFeedLikeRecord(client);
        this.post = new AppBskyFeedPostRecord(client);
        this.postgate = new AppBskyFeedPostgateRecord(client);
        this.repost = new AppBskyFeedRepostRecord(client);
        this.threadgate = new AppBskyFeedThreadgateRecord(client);
    }
    describeFeedGenerator(params, opts) {
        return this._client.call('app.bsky.feed.describeFeedGenerator', params, undefined, opts);
    }
    getActorFeeds(params, opts) {
        return this._client.call('app.bsky.feed.getActorFeeds', params, undefined, opts);
    }
    getActorLikes(params, opts) {
        return this._client
            .call('app.bsky.feed.getActorLikes', params, undefined, opts)
            .catch((e) => {
            throw AppBskyFeedGetActorLikes.toKnownErr(e);
        });
    }
    getAuthorFeed(params, opts) {
        return this._client
            .call('app.bsky.feed.getAuthorFeed', params, undefined, opts)
            .catch((e) => {
            throw AppBskyFeedGetAuthorFeed.toKnownErr(e);
        });
    }
    getFeed(params, opts) {
        return this._client
            .call('app.bsky.feed.getFeed', params, undefined, opts)
            .catch((e) => {
            throw AppBskyFeedGetFeed.toKnownErr(e);
        });
    }
    getFeedGenerator(params, opts) {
        return this._client.call('app.bsky.feed.getFeedGenerator', params, undefined, opts);
    }
    getFeedGenerators(params, opts) {
        return this._client.call('app.bsky.feed.getFeedGenerators', params, undefined, opts);
    }
    getFeedSkeleton(params, opts) {
        return this._client
            .call('app.bsky.feed.getFeedSkeleton', params, undefined, opts)
            .catch((e) => {
            throw AppBskyFeedGetFeedSkeleton.toKnownErr(e);
        });
    }
    getLikes(params, opts) {
        return this._client.call('app.bsky.feed.getLikes', params, undefined, opts);
    }
    getListFeed(params, opts) {
        return this._client
            .call('app.bsky.feed.getListFeed', params, undefined, opts)
            .catch((e) => {
            throw AppBskyFeedGetListFeed.toKnownErr(e);
        });
    }
    getPostThread(params, opts) {
        return this._client
            .call('app.bsky.feed.getPostThread', params, undefined, opts)
            .catch((e) => {
            throw AppBskyFeedGetPostThread.toKnownErr(e);
        });
    }
    getPosts(params, opts) {
        return this._client.call('app.bsky.feed.getPosts', params, undefined, opts);
    }
    getQuotes(params, opts) {
        return this._client.call('app.bsky.feed.getQuotes', params, undefined, opts);
    }
    getRepostedBy(params, opts) {
        return this._client.call('app.bsky.feed.getRepostedBy', params, undefined, opts);
    }
    getSuggestedFeeds(params, opts) {
        return this._client.call('app.bsky.feed.getSuggestedFeeds', params, undefined, opts);
    }
    getTimeline(params, opts) {
        return this._client.call('app.bsky.feed.getTimeline', params, undefined, opts);
    }
    searchPosts(params, opts) {
        return this._client
            .call('app.bsky.feed.searchPosts', params, undefined, opts)
            .catch((e) => {
            throw AppBskyFeedSearchPosts.toKnownErr(e);
        });
    }
    searchPostsV2(params, opts) {
        return this._client
            .call('app.bsky.feed.searchPostsV2', params, undefined, opts)
            .catch((e) => {
            throw AppBskyFeedSearchPostsV2.toKnownErr(e);
        });
    }
    sendInteractions(data, opts) {
        return this._client.call('app.bsky.feed.sendInteractions', opts?.qp, data, opts);
    }
}
export class AppBskyFeedGeneratorRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.feed.generator',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.feed.generator',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.feed.generator';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.feed.generator';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.feed.generator', ...params }, { headers });
    }
}
export class AppBskyFeedLikeRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.feed.like',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.feed.like',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.feed.like';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.feed.like';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.feed.like', ...params }, { headers });
    }
}
export class AppBskyFeedPostRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.feed.post',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.feed.post',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.feed.post';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.feed.post';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.feed.post', ...params }, { headers });
    }
}
export class AppBskyFeedPostgateRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.feed.postgate',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.feed.postgate',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.feed.postgate';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.feed.postgate';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.feed.postgate', ...params }, { headers });
    }
}
export class AppBskyFeedRepostRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.feed.repost',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.feed.repost',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.feed.repost';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.feed.repost';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.feed.repost', ...params }, { headers });
    }
}
export class AppBskyFeedThreadgateRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.feed.threadgate',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.feed.threadgate',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.feed.threadgate';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.feed.threadgate';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.feed.threadgate', ...params }, { headers });
    }
}
export class AppBskyGraphNS {
    constructor(client) {
        this._client = client;
        this.block = new AppBskyGraphBlockRecord(client);
        this.follow = new AppBskyGraphFollowRecord(client);
        this.list = new AppBskyGraphListRecord(client);
        this.listblock = new AppBskyGraphListblockRecord(client);
        this.listitem = new AppBskyGraphListitemRecord(client);
        this.starterpack = new AppBskyGraphStarterpackRecord(client);
        this.verification = new AppBskyGraphVerificationRecord(client);
    }
    getActorStarterPacks(params, opts) {
        return this._client.call('app.bsky.graph.getActorStarterPacks', params, undefined, opts);
    }
    getBlocks(params, opts) {
        return this._client.call('app.bsky.graph.getBlocks', params, undefined, opts);
    }
    getFollowers(params, opts) {
        return this._client.call('app.bsky.graph.getFollowers', params, undefined, opts);
    }
    getFollows(params, opts) {
        return this._client.call('app.bsky.graph.getFollows', params, undefined, opts);
    }
    getKnownFollowers(params, opts) {
        return this._client.call('app.bsky.graph.getKnownFollowers', params, undefined, opts);
    }
    getList(params, opts) {
        return this._client.call('app.bsky.graph.getList', params, undefined, opts);
    }
    getListBlocks(params, opts) {
        return this._client.call('app.bsky.graph.getListBlocks', params, undefined, opts);
    }
    getListMutes(params, opts) {
        return this._client.call('app.bsky.graph.getListMutes', params, undefined, opts);
    }
    getLists(params, opts) {
        return this._client.call('app.bsky.graph.getLists', params, undefined, opts);
    }
    getListsWithMembership(params, opts) {
        return this._client.call('app.bsky.graph.getListsWithMembership', params, undefined, opts);
    }
    getMutes(params, opts) {
        return this._client.call('app.bsky.graph.getMutes', params, undefined, opts);
    }
    getRelationships(params, opts) {
        return this._client
            .call('app.bsky.graph.getRelationships', params, undefined, opts)
            .catch((e) => {
            throw AppBskyGraphGetRelationships.toKnownErr(e);
        });
    }
    getStarterPack(params, opts) {
        return this._client.call('app.bsky.graph.getStarterPack', params, undefined, opts);
    }
    getStarterPacks(params, opts) {
        return this._client.call('app.bsky.graph.getStarterPacks', params, undefined, opts);
    }
    getStarterPacksWithMembership(params, opts) {
        return this._client.call('app.bsky.graph.getStarterPacksWithMembership', params, undefined, opts);
    }
    getSuggestedFollowsByActor(params, opts) {
        return this._client.call('app.bsky.graph.getSuggestedFollowsByActor', params, undefined, opts);
    }
    muteActor(data, opts) {
        return this._client.call('app.bsky.graph.muteActor', opts?.qp, data, opts);
    }
    muteActorList(data, opts) {
        return this._client.call('app.bsky.graph.muteActorList', opts?.qp, data, opts);
    }
    muteThread(data, opts) {
        return this._client.call('app.bsky.graph.muteThread', opts?.qp, data, opts);
    }
    searchStarterPacks(params, opts) {
        return this._client.call('app.bsky.graph.searchStarterPacks', params, undefined, opts);
    }
    unmuteActor(data, opts) {
        return this._client.call('app.bsky.graph.unmuteActor', opts?.qp, data, opts);
    }
    unmuteActorList(data, opts) {
        return this._client.call('app.bsky.graph.unmuteActorList', opts?.qp, data, opts);
    }
    unmuteThread(data, opts) {
        return this._client.call('app.bsky.graph.unmuteThread', opts?.qp, data, opts);
    }
}
export class AppBskyGraphBlockRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.graph.block',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.graph.block',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.graph.block';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.graph.block';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.graph.block', ...params }, { headers });
    }
}
export class AppBskyGraphFollowRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.graph.follow',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.graph.follow',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.graph.follow';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.graph.follow';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.graph.follow', ...params }, { headers });
    }
}
export class AppBskyGraphListRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.graph.list',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.graph.list',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.graph.list';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.graph.list';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.graph.list', ...params }, { headers });
    }
}
export class AppBskyGraphListblockRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.graph.listblock',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.graph.listblock',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.graph.listblock';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.graph.listblock';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.graph.listblock', ...params }, { headers });
    }
}
export class AppBskyGraphListitemRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.graph.listitem',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.graph.listitem',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.graph.listitem';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.graph.listitem';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.graph.listitem', ...params }, { headers });
    }
}
export class AppBskyGraphStarterpackRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.graph.starterpack',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.graph.starterpack',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.graph.starterpack';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.graph.starterpack';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.graph.starterpack', ...params }, { headers });
    }
}
export class AppBskyGraphVerificationRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.graph.verification',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.graph.verification',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.graph.verification';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.graph.verification';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.graph.verification', ...params }, { headers });
    }
}
export class AppBskyLabelerNS {
    constructor(client) {
        this._client = client;
        this.service = new AppBskyLabelerServiceRecord(client);
    }
    getServices(params, opts) {
        return this._client.call('app.bsky.labeler.getServices', params, undefined, opts);
    }
}
export class AppBskyLabelerServiceRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.labeler.service',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.labeler.service',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.labeler.service';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, {
            collection,
            rkey: 'self',
            ...params,
            record: { ...record, $type: collection },
        }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.labeler.service';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.labeler.service', ...params }, { headers });
    }
}
export class AppBskyNotificationNS {
    constructor(client) {
        this._client = client;
        this.declaration = new AppBskyNotificationDeclarationRecord(client);
    }
    getPreferences(params, opts) {
        return this._client.call('app.bsky.notification.getPreferences', params, undefined, opts);
    }
    getUnreadCount(params, opts) {
        return this._client.call('app.bsky.notification.getUnreadCount', params, undefined, opts);
    }
    listActivitySubscriptions(params, opts) {
        return this._client.call('app.bsky.notification.listActivitySubscriptions', params, undefined, opts);
    }
    listNotifications(params, opts) {
        return this._client.call('app.bsky.notification.listNotifications', params, undefined, opts);
    }
    putActivitySubscription(data, opts) {
        return this._client.call('app.bsky.notification.putActivitySubscription', opts?.qp, data, opts);
    }
    putPreferences(data, opts) {
        return this._client.call('app.bsky.notification.putPreferences', opts?.qp, data, opts);
    }
    putPreferencesV2(data, opts) {
        return this._client.call('app.bsky.notification.putPreferencesV2', opts?.qp, data, opts);
    }
    registerPush(data, opts) {
        return this._client.call('app.bsky.notification.registerPush', opts?.qp, data, opts);
    }
    unregisterPush(data, opts) {
        return this._client.call('app.bsky.notification.unregisterPush', opts?.qp, data, opts);
    }
    updateSeen(data, opts) {
        return this._client.call('app.bsky.notification.updateSeen', opts?.qp, data, opts);
    }
}
export class AppBskyNotificationDeclarationRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'app.bsky.notification.declaration',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'app.bsky.notification.declaration',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'app.bsky.notification.declaration';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, {
            collection,
            rkey: 'self',
            ...params,
            record: { ...record, $type: collection },
        }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'app.bsky.notification.declaration';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'app.bsky.notification.declaration', ...params }, { headers });
    }
}
export class AppBskyRichtextNS {
    constructor(client) {
        this._client = client;
    }
}
export class AppBskyUnspeccedNS {
    constructor(client) {
        this._client = client;
    }
    getAgeAssuranceState(params, opts) {
        return this._client.call('app.bsky.unspecced.getAgeAssuranceState', params, undefined, opts);
    }
    getConfig(params, opts) {
        return this._client.call('app.bsky.unspecced.getConfig', params, undefined, opts);
    }
    getOnboardingSuggestedStarterPacks(params, opts) {
        return this._client.call('app.bsky.unspecced.getOnboardingSuggestedStarterPacks', params, undefined, opts);
    }
    getOnboardingSuggestedStarterPacksSkeleton(params, opts) {
        return this._client.call('app.bsky.unspecced.getOnboardingSuggestedStarterPacksSkeleton', params, undefined, opts);
    }
    getOnboardingSuggestedUsersSkeleton(params, opts) {
        return this._client.call('app.bsky.unspecced.getOnboardingSuggestedUsersSkeleton', params, undefined, opts);
    }
    getPopularFeedGenerators(params, opts) {
        return this._client.call('app.bsky.unspecced.getPopularFeedGenerators', params, undefined, opts);
    }
    getPostThreadOtherV2(params, opts) {
        return this._client.call('app.bsky.unspecced.getPostThreadOtherV2', params, undefined, opts);
    }
    getPostThreadV2(params, opts) {
        return this._client.call('app.bsky.unspecced.getPostThreadV2', params, undefined, opts);
    }
    getSuggestedFeeds(params, opts) {
        return this._client.call('app.bsky.unspecced.getSuggestedFeeds', params, undefined, opts);
    }
    getSuggestedFeedsSkeleton(params, opts) {
        return this._client.call('app.bsky.unspecced.getSuggestedFeedsSkeleton', params, undefined, opts);
    }
    getSuggestedOnboardingUsers(params, opts) {
        return this._client.call('app.bsky.unspecced.getSuggestedOnboardingUsers', params, undefined, opts);
    }
    getSuggestedStarterPacks(params, opts) {
        return this._client.call('app.bsky.unspecced.getSuggestedStarterPacks', params, undefined, opts);
    }
    getSuggestedStarterPacksSkeleton(params, opts) {
        return this._client.call('app.bsky.unspecced.getSuggestedStarterPacksSkeleton', params, undefined, opts);
    }
    getSuggestedUsers(params, opts) {
        return this._client.call('app.bsky.unspecced.getSuggestedUsers', params, undefined, opts);
    }
    getSuggestedUsersForDiscover(params, opts) {
        return this._client.call('app.bsky.unspecced.getSuggestedUsersForDiscover', params, undefined, opts);
    }
    getSuggestedUsersForDiscoverSkeleton(params, opts) {
        return this._client.call('app.bsky.unspecced.getSuggestedUsersForDiscoverSkeleton', params, undefined, opts);
    }
    getSuggestedUsersForExplore(params, opts) {
        return this._client.call('app.bsky.unspecced.getSuggestedUsersForExplore', params, undefined, opts);
    }
    getSuggestedUsersForExploreSkeleton(params, opts) {
        return this._client.call('app.bsky.unspecced.getSuggestedUsersForExploreSkeleton', params, undefined, opts);
    }
    getSuggestedUsersForSeeMore(params, opts) {
        return this._client.call('app.bsky.unspecced.getSuggestedUsersForSeeMore', params, undefined, opts);
    }
    getSuggestedUsersForSeeMoreSkeleton(params, opts) {
        return this._client.call('app.bsky.unspecced.getSuggestedUsersForSeeMoreSkeleton', params, undefined, opts);
    }
    getSuggestedUsersSkeleton(params, opts) {
        return this._client.call('app.bsky.unspecced.getSuggestedUsersSkeleton', params, undefined, opts);
    }
    getSuggestionsSkeleton(params, opts) {
        return this._client.call('app.bsky.unspecced.getSuggestionsSkeleton', params, undefined, opts);
    }
    getTaggedSuggestions(params, opts) {
        return this._client.call('app.bsky.unspecced.getTaggedSuggestions', params, undefined, opts);
    }
    getTrendingTopics(params, opts) {
        return this._client.call('app.bsky.unspecced.getTrendingTopics', params, undefined, opts);
    }
    getTrends(params, opts) {
        return this._client.call('app.bsky.unspecced.getTrends', params, undefined, opts);
    }
    getTrendsSkeleton(params, opts) {
        return this._client.call('app.bsky.unspecced.getTrendsSkeleton', params, undefined, opts);
    }
    initAgeAssurance(data, opts) {
        return this._client
            .call('app.bsky.unspecced.initAgeAssurance', opts?.qp, data, opts)
            .catch((e) => {
            throw AppBskyUnspeccedInitAgeAssurance.toKnownErr(e);
        });
    }
    searchActorsSkeleton(params, opts) {
        return this._client
            .call('app.bsky.unspecced.searchActorsSkeleton', params, undefined, opts)
            .catch((e) => {
            throw AppBskyUnspeccedSearchActorsSkeleton.toKnownErr(e);
        });
    }
    searchPostsSkeleton(params, opts) {
        return this._client
            .call('app.bsky.unspecced.searchPostsSkeleton', params, undefined, opts)
            .catch((e) => {
            throw AppBskyUnspeccedSearchPostsSkeleton.toKnownErr(e);
        });
    }
    searchStarterPacksSkeleton(params, opts) {
        return this._client
            .call('app.bsky.unspecced.searchStarterPacksSkeleton', params, undefined, opts)
            .catch((e) => {
            throw AppBskyUnspeccedSearchStarterPacksSkeleton.toKnownErr(e);
        });
    }
}
export class AppBskyVideoNS {
    constructor(client) {
        this._client = client;
    }
    getJobStatus(params, opts) {
        return this._client.call('app.bsky.video.getJobStatus', params, undefined, opts);
    }
    getUploadLimits(params, opts) {
        return this._client.call('app.bsky.video.getUploadLimits', params, undefined, opts);
    }
    uploadVideo(data, opts) {
        return this._client.call('app.bsky.video.uploadVideo', opts?.qp, data, opts);
    }
}
export class ChatNS {
    constructor(client) {
        this._client = client;
        this.bsky = new ChatBskyNS(client);
    }
}
export class ChatBskyNS {
    constructor(client) {
        this._client = client;
        this.actor = new ChatBskyActorNS(client);
        this.convo = new ChatBskyConvoNS(client);
        this.embed = new ChatBskyEmbedNS(client);
        this.group = new ChatBskyGroupNS(client);
        this.moderation = new ChatBskyModerationNS(client);
        this.notification = new ChatBskyNotificationNS(client);
    }
}
export class ChatBskyActorNS {
    constructor(client) {
        this._client = client;
        this.declaration = new ChatBskyActorDeclarationRecord(client);
    }
    deleteAccount(data, opts) {
        return this._client.call('chat.bsky.actor.deleteAccount', opts?.qp, data, opts);
    }
    exportAccountData(params, opts) {
        return this._client.call('chat.bsky.actor.exportAccountData', params, undefined, opts);
    }
    getStatus(params, opts) {
        return this._client.call('chat.bsky.actor.getStatus', params, undefined, opts);
    }
}
export class ChatBskyActorDeclarationRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'chat.bsky.actor.declaration',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'chat.bsky.actor.declaration',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'chat.bsky.actor.declaration';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, {
            collection,
            rkey: 'self',
            ...params,
            record: { ...record, $type: collection },
        }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'chat.bsky.actor.declaration';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'chat.bsky.actor.declaration', ...params }, { headers });
    }
}
export class ChatBskyConvoNS {
    constructor(client) {
        this._client = client;
    }
    acceptConvo(data, opts) {
        return this._client
            .call('chat.bsky.convo.acceptConvo', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyConvoAcceptConvo.toKnownErr(e);
        });
    }
    addReaction(data, opts) {
        return this._client
            .call('chat.bsky.convo.addReaction', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyConvoAddReaction.toKnownErr(e);
        });
    }
    deleteMessageForSelf(data, opts) {
        return this._client
            .call('chat.bsky.convo.deleteMessageForSelf', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyConvoDeleteMessageForSelf.toKnownErr(e);
        });
    }
    getConvo(params, opts) {
        return this._client
            .call('chat.bsky.convo.getConvo', params, undefined, opts)
            .catch((e) => {
            throw ChatBskyConvoGetConvo.toKnownErr(e);
        });
    }
    getConvoAvailability(params, opts) {
        return this._client.call('chat.bsky.convo.getConvoAvailability', params, undefined, opts);
    }
    getConvoForMembers(params, opts) {
        return this._client
            .call('chat.bsky.convo.getConvoForMembers', params, undefined, opts)
            .catch((e) => {
            throw ChatBskyConvoGetConvoForMembers.toKnownErr(e);
        });
    }
    getConvoMembers(params, opts) {
        return this._client
            .call('chat.bsky.convo.getConvoMembers', params, undefined, opts)
            .catch((e) => {
            throw ChatBskyConvoGetConvoMembers.toKnownErr(e);
        });
    }
    getLog(params, opts) {
        return this._client.call('chat.bsky.convo.getLog', params, undefined, opts);
    }
    getMessages(params, opts) {
        return this._client
            .call('chat.bsky.convo.getMessages', params, undefined, opts)
            .catch((e) => {
            throw ChatBskyConvoGetMessages.toKnownErr(e);
        });
    }
    getUnreadCounts(params, opts) {
        return this._client.call('chat.bsky.convo.getUnreadCounts', params, undefined, opts);
    }
    leaveConvo(data, opts) {
        return this._client
            .call('chat.bsky.convo.leaveConvo', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyConvoLeaveConvo.toKnownErr(e);
        });
    }
    listConvoRequests(params, opts) {
        return this._client.call('chat.bsky.convo.listConvoRequests', params, undefined, opts);
    }
    listConvos(params, opts) {
        return this._client.call('chat.bsky.convo.listConvos', params, undefined, opts);
    }
    lockConvo(data, opts) {
        return this._client
            .call('chat.bsky.convo.lockConvo', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyConvoLockConvo.toKnownErr(e);
        });
    }
    muteConvo(data, opts) {
        return this._client
            .call('chat.bsky.convo.muteConvo', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyConvoMuteConvo.toKnownErr(e);
        });
    }
    removeReaction(data, opts) {
        return this._client
            .call('chat.bsky.convo.removeReaction', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyConvoRemoveReaction.toKnownErr(e);
        });
    }
    sendMessage(data, opts) {
        return this._client
            .call('chat.bsky.convo.sendMessage', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyConvoSendMessage.toKnownErr(e);
        });
    }
    sendMessageBatch(data, opts) {
        return this._client
            .call('chat.bsky.convo.sendMessageBatch', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyConvoSendMessageBatch.toKnownErr(e);
        });
    }
    unlockConvo(data, opts) {
        return this._client
            .call('chat.bsky.convo.unlockConvo', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyConvoUnlockConvo.toKnownErr(e);
        });
    }
    unmuteConvo(data, opts) {
        return this._client
            .call('chat.bsky.convo.unmuteConvo', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyConvoUnmuteConvo.toKnownErr(e);
        });
    }
    updateAllRead(data, opts) {
        return this._client.call('chat.bsky.convo.updateAllRead', opts?.qp, data, opts);
    }
    updateRead(data, opts) {
        return this._client
            .call('chat.bsky.convo.updateRead', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyConvoUpdateRead.toKnownErr(e);
        });
    }
}
export class ChatBskyEmbedNS {
    constructor(client) {
        this._client = client;
    }
}
export class ChatBskyGroupNS {
    constructor(client) {
        this._client = client;
    }
    addMembers(data, opts) {
        return this._client
            .call('chat.bsky.group.addMembers', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyGroupAddMembers.toKnownErr(e);
        });
    }
    approveJoinRequest(data, opts) {
        return this._client
            .call('chat.bsky.group.approveJoinRequest', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyGroupApproveJoinRequest.toKnownErr(e);
        });
    }
    createGroup(data, opts) {
        return this._client
            .call('chat.bsky.group.createGroup', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyGroupCreateGroup.toKnownErr(e);
        });
    }
    createJoinLink(data, opts) {
        return this._client
            .call('chat.bsky.group.createJoinLink', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyGroupCreateJoinLink.toKnownErr(e);
        });
    }
    disableJoinLink(data, opts) {
        return this._client
            .call('chat.bsky.group.disableJoinLink', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyGroupDisableJoinLink.toKnownErr(e);
        });
    }
    editGroup(data, opts) {
        return this._client
            .call('chat.bsky.group.editGroup', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyGroupEditGroup.toKnownErr(e);
        });
    }
    editJoinLink(data, opts) {
        return this._client
            .call('chat.bsky.group.editJoinLink', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyGroupEditJoinLink.toKnownErr(e);
        });
    }
    enableJoinLink(data, opts) {
        return this._client
            .call('chat.bsky.group.enableJoinLink', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyGroupEnableJoinLink.toKnownErr(e);
        });
    }
    getJoinLinkPreviews(params, opts) {
        return this._client.call('chat.bsky.group.getJoinLinkPreviews', params, undefined, opts);
    }
    listJoinRequests(params, opts) {
        return this._client
            .call('chat.bsky.group.listJoinRequests', params, undefined, opts)
            .catch((e) => {
            throw ChatBskyGroupListJoinRequests.toKnownErr(e);
        });
    }
    listMutualGroups(params, opts) {
        return this._client.call('chat.bsky.group.listMutualGroups', params, undefined, opts);
    }
    rejectJoinRequest(data, opts) {
        return this._client
            .call('chat.bsky.group.rejectJoinRequest', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyGroupRejectJoinRequest.toKnownErr(e);
        });
    }
    removeMembers(data, opts) {
        return this._client
            .call('chat.bsky.group.removeMembers', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyGroupRemoveMembers.toKnownErr(e);
        });
    }
    requestJoin(data, opts) {
        return this._client
            .call('chat.bsky.group.requestJoin', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyGroupRequestJoin.toKnownErr(e);
        });
    }
    updateJoinRequestsRead(data, opts) {
        return this._client
            .call('chat.bsky.group.updateJoinRequestsRead', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyGroupUpdateJoinRequestsRead.toKnownErr(e);
        });
    }
    withdrawJoinRequest(data, opts) {
        return this._client
            .call('chat.bsky.group.withdrawJoinRequest', opts?.qp, data, opts)
            .catch((e) => {
            throw ChatBskyGroupWithdrawJoinRequest.toKnownErr(e);
        });
    }
}
export class ChatBskyModerationNS {
    constructor(client) {
        this._client = client;
    }
    getActorMetadata(params, opts) {
        return this._client.call('chat.bsky.moderation.getActorMetadata', params, undefined, opts);
    }
    getConvo(params, opts) {
        return this._client
            .call('chat.bsky.moderation.getConvo', params, undefined, opts)
            .catch((e) => {
            throw ChatBskyModerationGetConvo.toKnownErr(e);
        });
    }
    getConvoMembers(params, opts) {
        return this._client
            .call('chat.bsky.moderation.getConvoMembers', params, undefined, opts)
            .catch((e) => {
            throw ChatBskyModerationGetConvoMembers.toKnownErr(e);
        });
    }
    getConvos(params, opts) {
        return this._client.call('chat.bsky.moderation.getConvos', params, undefined, opts);
    }
    getMessageContext(params, opts) {
        return this._client.call('chat.bsky.moderation.getMessageContext', params, undefined, opts);
    }
    updateActorAccess(data, opts) {
        return this._client.call('chat.bsky.moderation.updateActorAccess', opts?.qp, data, opts);
    }
}
export class ChatBskyNotificationNS {
    constructor(client) {
        this._client = client;
    }
    getPreferences(params, opts) {
        return this._client.call('chat.bsky.notification.getPreferences', params, undefined, opts);
    }
    putPreferences(data, opts) {
        return this._client.call('chat.bsky.notification.putPreferences', opts?.qp, data, opts);
    }
}
export class ComNS {
    constructor(client) {
        this._client = client;
        this.atproto = new ComAtprotoNS(client);
        this.germnetwork = new ComGermnetworkNS(client);
    }
}
export class ComAtprotoNS {
    constructor(client) {
        this._client = client;
        this.admin = new ComAtprotoAdminNS(client);
        this.identity = new ComAtprotoIdentityNS(client);
        this.label = new ComAtprotoLabelNS(client);
        this.lexicon = new ComAtprotoLexiconNS(client);
        this.moderation = new ComAtprotoModerationNS(client);
        this.repo = new ComAtprotoRepoNS(client);
        this.server = new ComAtprotoServerNS(client);
        this.sync = new ComAtprotoSyncNS(client);
        this.temp = new ComAtprotoTempNS(client);
    }
}
export class ComAtprotoAdminNS {
    constructor(client) {
        this._client = client;
    }
    deleteAccount(data, opts) {
        return this._client.call('com.atproto.admin.deleteAccount', opts?.qp, data, opts);
    }
    disableAccountInvites(data, opts) {
        return this._client.call('com.atproto.admin.disableAccountInvites', opts?.qp, data, opts);
    }
    disableInviteCodes(data, opts) {
        return this._client.call('com.atproto.admin.disableInviteCodes', opts?.qp, data, opts);
    }
    enableAccountInvites(data, opts) {
        return this._client.call('com.atproto.admin.enableAccountInvites', opts?.qp, data, opts);
    }
    getAccountInfo(params, opts) {
        return this._client.call('com.atproto.admin.getAccountInfo', params, undefined, opts);
    }
    getAccountInfos(params, opts) {
        return this._client.call('com.atproto.admin.getAccountInfos', params, undefined, opts);
    }
    getInviteCodes(params, opts) {
        return this._client.call('com.atproto.admin.getInviteCodes', params, undefined, opts);
    }
    getSubjectStatus(params, opts) {
        return this._client.call('com.atproto.admin.getSubjectStatus', params, undefined, opts);
    }
    searchAccounts(params, opts) {
        return this._client.call('com.atproto.admin.searchAccounts', params, undefined, opts);
    }
    sendEmail(data, opts) {
        return this._client.call('com.atproto.admin.sendEmail', opts?.qp, data, opts);
    }
    updateAccountEmail(data, opts) {
        return this._client.call('com.atproto.admin.updateAccountEmail', opts?.qp, data, opts);
    }
    updateAccountHandle(data, opts) {
        return this._client.call('com.atproto.admin.updateAccountHandle', opts?.qp, data, opts);
    }
    updateAccountPassword(data, opts) {
        return this._client.call('com.atproto.admin.updateAccountPassword', opts?.qp, data, opts);
    }
    updateAccountSigningKey(data, opts) {
        return this._client.call('com.atproto.admin.updateAccountSigningKey', opts?.qp, data, opts);
    }
    updateSubjectStatus(data, opts) {
        return this._client.call('com.atproto.admin.updateSubjectStatus', opts?.qp, data, opts);
    }
}
export class ComAtprotoIdentityNS {
    constructor(client) {
        this._client = client;
    }
    getRecommendedDidCredentials(params, opts) {
        return this._client.call('com.atproto.identity.getRecommendedDidCredentials', params, undefined, opts);
    }
    refreshIdentity(data, opts) {
        return this._client
            .call('com.atproto.identity.refreshIdentity', opts?.qp, data, opts)
            .catch((e) => {
            throw ComAtprotoIdentityRefreshIdentity.toKnownErr(e);
        });
    }
    requestPlcOperationSignature(data, opts) {
        return this._client.call('com.atproto.identity.requestPlcOperationSignature', opts?.qp, data, opts);
    }
    resolveDid(params, opts) {
        return this._client
            .call('com.atproto.identity.resolveDid', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoIdentityResolveDid.toKnownErr(e);
        });
    }
    resolveHandle(params, opts) {
        return this._client
            .call('com.atproto.identity.resolveHandle', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoIdentityResolveHandle.toKnownErr(e);
        });
    }
    resolveIdentity(params, opts) {
        return this._client
            .call('com.atproto.identity.resolveIdentity', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoIdentityResolveIdentity.toKnownErr(e);
        });
    }
    signPlcOperation(data, opts) {
        return this._client.call('com.atproto.identity.signPlcOperation', opts?.qp, data, opts);
    }
    submitPlcOperation(data, opts) {
        return this._client.call('com.atproto.identity.submitPlcOperation', opts?.qp, data, opts);
    }
    updateHandle(data, opts) {
        return this._client.call('com.atproto.identity.updateHandle', opts?.qp, data, opts);
    }
}
export class ComAtprotoLabelNS {
    constructor(client) {
        this._client = client;
    }
    queryLabels(params, opts) {
        return this._client.call('com.atproto.label.queryLabels', params, undefined, opts);
    }
}
export class ComAtprotoLexiconNS {
    constructor(client) {
        this._client = client;
        this.schema = new ComAtprotoLexiconSchemaRecord(client);
    }
    resolveLexicon(params, opts) {
        return this._client
            .call('com.atproto.lexicon.resolveLexicon', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoLexiconResolveLexicon.toKnownErr(e);
        });
    }
}
export class ComAtprotoLexiconSchemaRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'com.atproto.lexicon.schema',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'com.atproto.lexicon.schema',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'com.atproto.lexicon.schema';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'com.atproto.lexicon.schema';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'com.atproto.lexicon.schema', ...params }, { headers });
    }
}
export class ComAtprotoModerationNS {
    constructor(client) {
        this._client = client;
    }
    createReport(data, opts) {
        return this._client.call('com.atproto.moderation.createReport', opts?.qp, data, opts);
    }
}
export class ComAtprotoRepoNS {
    constructor(client) {
        this._client = client;
    }
    applyWrites(data, opts) {
        return this._client
            .call('com.atproto.repo.applyWrites', opts?.qp, data, opts)
            .catch((e) => {
            throw ComAtprotoRepoApplyWrites.toKnownErr(e);
        });
    }
    createRecord(data, opts) {
        return this._client
            .call('com.atproto.repo.createRecord', opts?.qp, data, opts)
            .catch((e) => {
            throw ComAtprotoRepoCreateRecord.toKnownErr(e);
        });
    }
    deleteRecord(data, opts) {
        return this._client
            .call('com.atproto.repo.deleteRecord', opts?.qp, data, opts)
            .catch((e) => {
            throw ComAtprotoRepoDeleteRecord.toKnownErr(e);
        });
    }
    describeRepo(params, opts) {
        return this._client.call('com.atproto.repo.describeRepo', params, undefined, opts);
    }
    getRecord(params, opts) {
        return this._client
            .call('com.atproto.repo.getRecord', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoRepoGetRecord.toKnownErr(e);
        });
    }
    importRepo(data, opts) {
        return this._client.call('com.atproto.repo.importRepo', opts?.qp, data, opts);
    }
    listMissingBlobs(params, opts) {
        return this._client.call('com.atproto.repo.listMissingBlobs', params, undefined, opts);
    }
    listRecords(params, opts) {
        return this._client.call('com.atproto.repo.listRecords', params, undefined, opts);
    }
    putRecord(data, opts) {
        return this._client
            .call('com.atproto.repo.putRecord', opts?.qp, data, opts)
            .catch((e) => {
            throw ComAtprotoRepoPutRecord.toKnownErr(e);
        });
    }
    uploadBlob(data, opts) {
        return this._client.call('com.atproto.repo.uploadBlob', opts?.qp, data, opts);
    }
}
export class ComAtprotoServerNS {
    constructor(client) {
        this._client = client;
    }
    activateAccount(data, opts) {
        return this._client.call('com.atproto.server.activateAccount', opts?.qp, data, opts);
    }
    checkAccountStatus(params, opts) {
        return this._client.call('com.atproto.server.checkAccountStatus', params, undefined, opts);
    }
    confirmEmail(data, opts) {
        return this._client
            .call('com.atproto.server.confirmEmail', opts?.qp, data, opts)
            .catch((e) => {
            throw ComAtprotoServerConfirmEmail.toKnownErr(e);
        });
    }
    createAccount(data, opts) {
        return this._client
            .call('com.atproto.server.createAccount', opts?.qp, data, opts)
            .catch((e) => {
            throw ComAtprotoServerCreateAccount.toKnownErr(e);
        });
    }
    createAppPassword(data, opts) {
        return this._client
            .call('com.atproto.server.createAppPassword', opts?.qp, data, opts)
            .catch((e) => {
            throw ComAtprotoServerCreateAppPassword.toKnownErr(e);
        });
    }
    createInviteCode(data, opts) {
        return this._client.call('com.atproto.server.createInviteCode', opts?.qp, data, opts);
    }
    createInviteCodes(data, opts) {
        return this._client.call('com.atproto.server.createInviteCodes', opts?.qp, data, opts);
    }
    createSession(data, opts) {
        return this._client
            .call('com.atproto.server.createSession', opts?.qp, data, opts)
            .catch((e) => {
            throw ComAtprotoServerCreateSession.toKnownErr(e);
        });
    }
    deactivateAccount(data, opts) {
        return this._client.call('com.atproto.server.deactivateAccount', opts?.qp, data, opts);
    }
    deleteAccount(data, opts) {
        return this._client
            .call('com.atproto.server.deleteAccount', opts?.qp, data, opts)
            .catch((e) => {
            throw ComAtprotoServerDeleteAccount.toKnownErr(e);
        });
    }
    deleteSession(data, opts) {
        return this._client
            .call('com.atproto.server.deleteSession', opts?.qp, data, opts)
            .catch((e) => {
            throw ComAtprotoServerDeleteSession.toKnownErr(e);
        });
    }
    describeServer(params, opts) {
        return this._client.call('com.atproto.server.describeServer', params, undefined, opts);
    }
    getAccountInviteCodes(params, opts) {
        return this._client
            .call('com.atproto.server.getAccountInviteCodes', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoServerGetAccountInviteCodes.toKnownErr(e);
        });
    }
    getServiceAuth(params, opts) {
        return this._client
            .call('com.atproto.server.getServiceAuth', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoServerGetServiceAuth.toKnownErr(e);
        });
    }
    getSession(params, opts) {
        return this._client.call('com.atproto.server.getSession', params, undefined, opts);
    }
    listAppPasswords(params, opts) {
        return this._client
            .call('com.atproto.server.listAppPasswords', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoServerListAppPasswords.toKnownErr(e);
        });
    }
    refreshSession(data, opts) {
        return this._client
            .call('com.atproto.server.refreshSession', opts?.qp, data, opts)
            .catch((e) => {
            throw ComAtprotoServerRefreshSession.toKnownErr(e);
        });
    }
    requestAccountDelete(data, opts) {
        return this._client.call('com.atproto.server.requestAccountDelete', opts?.qp, data, opts);
    }
    requestEmailConfirmation(data, opts) {
        return this._client.call('com.atproto.server.requestEmailConfirmation', opts?.qp, data, opts);
    }
    requestEmailUpdate(data, opts) {
        return this._client.call('com.atproto.server.requestEmailUpdate', opts?.qp, data, opts);
    }
    requestPasswordReset(data, opts) {
        return this._client.call('com.atproto.server.requestPasswordReset', opts?.qp, data, opts);
    }
    reserveSigningKey(data, opts) {
        return this._client.call('com.atproto.server.reserveSigningKey', opts?.qp, data, opts);
    }
    resetPassword(data, opts) {
        return this._client
            .call('com.atproto.server.resetPassword', opts?.qp, data, opts)
            .catch((e) => {
            throw ComAtprotoServerResetPassword.toKnownErr(e);
        });
    }
    revokeAppPassword(data, opts) {
        return this._client.call('com.atproto.server.revokeAppPassword', opts?.qp, data, opts);
    }
    updateEmail(data, opts) {
        return this._client
            .call('com.atproto.server.updateEmail', opts?.qp, data, opts)
            .catch((e) => {
            throw ComAtprotoServerUpdateEmail.toKnownErr(e);
        });
    }
}
export class ComAtprotoSyncNS {
    constructor(client) {
        this._client = client;
    }
    getBlob(params, opts) {
        return this._client
            .call('com.atproto.sync.getBlob', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoSyncGetBlob.toKnownErr(e);
        });
    }
    getBlocks(params, opts) {
        return this._client
            .call('com.atproto.sync.getBlocks', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoSyncGetBlocks.toKnownErr(e);
        });
    }
    getCheckout(params, opts) {
        return this._client.call('com.atproto.sync.getCheckout', params, undefined, opts);
    }
    getHead(params, opts) {
        return this._client
            .call('com.atproto.sync.getHead', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoSyncGetHead.toKnownErr(e);
        });
    }
    getHostStatus(params, opts) {
        return this._client
            .call('com.atproto.sync.getHostStatus', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoSyncGetHostStatus.toKnownErr(e);
        });
    }
    getLatestCommit(params, opts) {
        return this._client
            .call('com.atproto.sync.getLatestCommit', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoSyncGetLatestCommit.toKnownErr(e);
        });
    }
    getRecord(params, opts) {
        return this._client
            .call('com.atproto.sync.getRecord', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoSyncGetRecord.toKnownErr(e);
        });
    }
    getRepo(params, opts) {
        return this._client
            .call('com.atproto.sync.getRepo', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoSyncGetRepo.toKnownErr(e);
        });
    }
    getRepoStatus(params, opts) {
        return this._client
            .call('com.atproto.sync.getRepoStatus', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoSyncGetRepoStatus.toKnownErr(e);
        });
    }
    listBlobs(params, opts) {
        return this._client
            .call('com.atproto.sync.listBlobs', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoSyncListBlobs.toKnownErr(e);
        });
    }
    listHosts(params, opts) {
        return this._client.call('com.atproto.sync.listHosts', params, undefined, opts);
    }
    listRepos(params, opts) {
        return this._client.call('com.atproto.sync.listRepos', params, undefined, opts);
    }
    listReposByCollection(params, opts) {
        return this._client.call('com.atproto.sync.listReposByCollection', params, undefined, opts);
    }
    notifyOfUpdate(data, opts) {
        return this._client.call('com.atproto.sync.notifyOfUpdate', opts?.qp, data, opts);
    }
    requestCrawl(data, opts) {
        return this._client
            .call('com.atproto.sync.requestCrawl', opts?.qp, data, opts)
            .catch((e) => {
            throw ComAtprotoSyncRequestCrawl.toKnownErr(e);
        });
    }
}
export class ComAtprotoTempNS {
    constructor(client) {
        this._client = client;
    }
    addReservedHandle(data, opts) {
        return this._client.call('com.atproto.temp.addReservedHandle', opts?.qp, data, opts);
    }
    checkHandleAvailability(params, opts) {
        return this._client
            .call('com.atproto.temp.checkHandleAvailability', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoTempCheckHandleAvailability.toKnownErr(e);
        });
    }
    checkSignupQueue(params, opts) {
        return this._client.call('com.atproto.temp.checkSignupQueue', params, undefined, opts);
    }
    dereferenceScope(params, opts) {
        return this._client
            .call('com.atproto.temp.dereferenceScope', params, undefined, opts)
            .catch((e) => {
            throw ComAtprotoTempDereferenceScope.toKnownErr(e);
        });
    }
    fetchLabels(params, opts) {
        return this._client.call('com.atproto.temp.fetchLabels', params, undefined, opts);
    }
    requestPhoneVerification(data, opts) {
        return this._client.call('com.atproto.temp.requestPhoneVerification', opts?.qp, data, opts);
    }
    revokeAccountCredentials(data, opts) {
        return this._client.call('com.atproto.temp.revokeAccountCredentials', opts?.qp, data, opts);
    }
}
export class ComGermnetworkNS {
    constructor(client) {
        this._client = client;
        this.declaration = new ComGermnetworkDeclarationRecord(client);
    }
}
export class ComGermnetworkDeclarationRecord {
    constructor(client) {
        this._client = client;
    }
    async list(params) {
        const res = await this._client.call('com.atproto.repo.listRecords', {
            collection: 'com.germnetwork.declaration',
            ...params,
        });
        return res.data;
    }
    async get(params) {
        const res = await this._client.call('com.atproto.repo.getRecord', {
            collection: 'com.germnetwork.declaration',
            ...params,
        });
        return res.data;
    }
    async create(params, record, headers) {
        const collection = 'com.germnetwork.declaration';
        const res = await this._client.call('com.atproto.repo.createRecord', undefined, {
            collection,
            rkey: 'self',
            ...params,
            record: { ...record, $type: collection },
        }, { encoding: 'application/json', headers });
        return res.data;
    }
    async put(params, record, headers) {
        const collection = 'com.germnetwork.declaration';
        const res = await this._client.call('com.atproto.repo.putRecord', undefined, { collection, ...params, record: { ...record, $type: collection } }, { encoding: 'application/json', headers });
        return res.data;
    }
    async delete(params, headers) {
        await this._client.call('com.atproto.repo.deleteRecord', undefined, { collection: 'com.germnetwork.declaration', ...params }, { headers });
    }
}
export class ToolsNS {
    constructor(client) {
        this._client = client;
        this.ozone = new ToolsOzoneNS(client);
    }
}
export class ToolsOzoneNS {
    constructor(client) {
        this._client = client;
        this.communication = new ToolsOzoneCommunicationNS(client);
        this.hosting = new ToolsOzoneHostingNS(client);
        this.moderation = new ToolsOzoneModerationNS(client);
        this.queue = new ToolsOzoneQueueNS(client);
        this.report = new ToolsOzoneReportNS(client);
        this.safelink = new ToolsOzoneSafelinkNS(client);
        this.server = new ToolsOzoneServerNS(client);
        this.set = new ToolsOzoneSetNS(client);
        this.setting = new ToolsOzoneSettingNS(client);
        this.signature = new ToolsOzoneSignatureNS(client);
        this.team = new ToolsOzoneTeamNS(client);
        this.verification = new ToolsOzoneVerificationNS(client);
    }
}
export class ToolsOzoneCommunicationNS {
    constructor(client) {
        this._client = client;
    }
    createTemplate(data, opts) {
        return this._client
            .call('tools.ozone.communication.createTemplate', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneCommunicationCreateTemplate.toKnownErr(e);
        });
    }
    deleteTemplate(data, opts) {
        return this._client.call('tools.ozone.communication.deleteTemplate', opts?.qp, data, opts);
    }
    listTemplates(params, opts) {
        return this._client.call('tools.ozone.communication.listTemplates', params, undefined, opts);
    }
    updateTemplate(data, opts) {
        return this._client
            .call('tools.ozone.communication.updateTemplate', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneCommunicationUpdateTemplate.toKnownErr(e);
        });
    }
}
export class ToolsOzoneHostingNS {
    constructor(client) {
        this._client = client;
    }
    getAccountHistory(params, opts) {
        return this._client.call('tools.ozone.hosting.getAccountHistory', params, undefined, opts);
    }
}
export class ToolsOzoneModerationNS {
    constructor(client) {
        this._client = client;
    }
    cancelScheduledActions(data, opts) {
        return this._client.call('tools.ozone.moderation.cancelScheduledActions', opts?.qp, data, opts);
    }
    emitEvent(data, opts) {
        return this._client
            .call('tools.ozone.moderation.emitEvent', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneModerationEmitEvent.toKnownErr(e);
        });
    }
    getAccountTimeline(params, opts) {
        return this._client
            .call('tools.ozone.moderation.getAccountTimeline', params, undefined, opts)
            .catch((e) => {
            throw ToolsOzoneModerationGetAccountTimeline.toKnownErr(e);
        });
    }
    getEvent(params, opts) {
        return this._client.call('tools.ozone.moderation.getEvent', params, undefined, opts);
    }
    getRecord(params, opts) {
        return this._client
            .call('tools.ozone.moderation.getRecord', params, undefined, opts)
            .catch((e) => {
            throw ToolsOzoneModerationGetRecord.toKnownErr(e);
        });
    }
    getRecords(params, opts) {
        return this._client.call('tools.ozone.moderation.getRecords', params, undefined, opts);
    }
    getRepo(params, opts) {
        return this._client
            .call('tools.ozone.moderation.getRepo', params, undefined, opts)
            .catch((e) => {
            throw ToolsOzoneModerationGetRepo.toKnownErr(e);
        });
    }
    getReporterStats(params, opts) {
        return this._client.call('tools.ozone.moderation.getReporterStats', params, undefined, opts);
    }
    getRepos(params, opts) {
        return this._client.call('tools.ozone.moderation.getRepos', params, undefined, opts);
    }
    getSubjects(params, opts) {
        return this._client.call('tools.ozone.moderation.getSubjects', params, undefined, opts);
    }
    listScheduledActions(data, opts) {
        return this._client.call('tools.ozone.moderation.listScheduledActions', opts?.qp, data, opts);
    }
    queryEvents(params, opts) {
        return this._client.call('tools.ozone.moderation.queryEvents', params, undefined, opts);
    }
    queryStatuses(params, opts) {
        return this._client.call('tools.ozone.moderation.queryStatuses', params, undefined, opts);
    }
    scheduleAction(data, opts) {
        return this._client.call('tools.ozone.moderation.scheduleAction', opts?.qp, data, opts);
    }
    searchRepos(params, opts) {
        return this._client.call('tools.ozone.moderation.searchRepos', params, undefined, opts);
    }
}
export class ToolsOzoneQueueNS {
    constructor(client) {
        this._client = client;
    }
    assignModerator(data, opts) {
        return this._client
            .call('tools.ozone.queue.assignModerator', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneQueueAssignModerator.toKnownErr(e);
        });
    }
    createQueue(data, opts) {
        return this._client
            .call('tools.ozone.queue.createQueue', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneQueueCreateQueue.toKnownErr(e);
        });
    }
    deleteQueue(data, opts) {
        return this._client.call('tools.ozone.queue.deleteQueue', opts?.qp, data, opts);
    }
    getAssignments(params, opts) {
        return this._client.call('tools.ozone.queue.getAssignments', params, undefined, opts);
    }
    listQueues(params, opts) {
        return this._client.call('tools.ozone.queue.listQueues', params, undefined, opts);
    }
    routeReports(data, opts) {
        return this._client
            .call('tools.ozone.queue.routeReports', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneQueueRouteReports.toKnownErr(e);
        });
    }
    unassignModerator(data, opts) {
        return this._client
            .call('tools.ozone.queue.unassignModerator', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneQueueUnassignModerator.toKnownErr(e);
        });
    }
    updateQueue(data, opts) {
        return this._client.call('tools.ozone.queue.updateQueue', opts?.qp, data, opts);
    }
}
export class ToolsOzoneReportNS {
    constructor(client) {
        this._client = client;
    }
    assignModerator(data, opts) {
        return this._client
            .call('tools.ozone.report.assignModerator', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneReportAssignModerator.toKnownErr(e);
        });
    }
    createActivity(data, opts) {
        return this._client
            .call('tools.ozone.report.createActivity', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneReportCreateActivity.toKnownErr(e);
        });
    }
    getAssignments(params, opts) {
        return this._client.call('tools.ozone.report.getAssignments', params, undefined, opts);
    }
    getHistoricalStats(params, opts) {
        return this._client.call('tools.ozone.report.getHistoricalStats', params, undefined, opts);
    }
    getLatestReport(params, opts) {
        return this._client
            .call('tools.ozone.report.getLatestReport', params, undefined, opts)
            .catch((e) => {
            throw ToolsOzoneReportGetLatestReport.toKnownErr(e);
        });
    }
    getLiveStats(params, opts) {
        return this._client.call('tools.ozone.report.getLiveStats', params, undefined, opts);
    }
    getReport(params, opts) {
        return this._client
            .call('tools.ozone.report.getReport', params, undefined, opts)
            .catch((e) => {
            throw ToolsOzoneReportGetReport.toKnownErr(e);
        });
    }
    listActivities(params, opts) {
        return this._client.call('tools.ozone.report.listActivities', params, undefined, opts);
    }
    queryActivities(params, opts) {
        return this._client.call('tools.ozone.report.queryActivities', params, undefined, opts);
    }
    queryReports(params, opts) {
        return this._client.call('tools.ozone.report.queryReports', params, undefined, opts);
    }
    reassignQueue(data, opts) {
        return this._client
            .call('tools.ozone.report.reassignQueue', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneReportReassignQueue.toKnownErr(e);
        });
    }
    refreshStats(data, opts) {
        return this._client.call('tools.ozone.report.refreshStats', opts?.qp, data, opts);
    }
    unassignModerator(data, opts) {
        return this._client
            .call('tools.ozone.report.unassignModerator', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneReportUnassignModerator.toKnownErr(e);
        });
    }
}
export class ToolsOzoneSafelinkNS {
    constructor(client) {
        this._client = client;
    }
    addRule(data, opts) {
        return this._client
            .call('tools.ozone.safelink.addRule', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneSafelinkAddRule.toKnownErr(e);
        });
    }
    queryEvents(data, opts) {
        return this._client.call('tools.ozone.safelink.queryEvents', opts?.qp, data, opts);
    }
    queryRules(data, opts) {
        return this._client.call('tools.ozone.safelink.queryRules', opts?.qp, data, opts);
    }
    removeRule(data, opts) {
        return this._client
            .call('tools.ozone.safelink.removeRule', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneSafelinkRemoveRule.toKnownErr(e);
        });
    }
    updateRule(data, opts) {
        return this._client
            .call('tools.ozone.safelink.updateRule', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneSafelinkUpdateRule.toKnownErr(e);
        });
    }
}
export class ToolsOzoneServerNS {
    constructor(client) {
        this._client = client;
    }
    getConfig(params, opts) {
        return this._client.call('tools.ozone.server.getConfig', params, undefined, opts);
    }
}
export class ToolsOzoneSetNS {
    constructor(client) {
        this._client = client;
    }
    addValues(data, opts) {
        return this._client.call('tools.ozone.set.addValues', opts?.qp, data, opts);
    }
    deleteSet(data, opts) {
        return this._client
            .call('tools.ozone.set.deleteSet', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneSetDeleteSet.toKnownErr(e);
        });
    }
    deleteValues(data, opts) {
        return this._client
            .call('tools.ozone.set.deleteValues', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneSetDeleteValues.toKnownErr(e);
        });
    }
    getValues(params, opts) {
        return this._client
            .call('tools.ozone.set.getValues', params, undefined, opts)
            .catch((e) => {
            throw ToolsOzoneSetGetValues.toKnownErr(e);
        });
    }
    querySets(params, opts) {
        return this._client.call('tools.ozone.set.querySets', params, undefined, opts);
    }
    upsertSet(data, opts) {
        return this._client.call('tools.ozone.set.upsertSet', opts?.qp, data, opts);
    }
}
export class ToolsOzoneSettingNS {
    constructor(client) {
        this._client = client;
    }
    listOptions(params, opts) {
        return this._client.call('tools.ozone.setting.listOptions', params, undefined, opts);
    }
    removeOptions(data, opts) {
        return this._client.call('tools.ozone.setting.removeOptions', opts?.qp, data, opts);
    }
    upsertOption(data, opts) {
        return this._client.call('tools.ozone.setting.upsertOption', opts?.qp, data, opts);
    }
}
export class ToolsOzoneSignatureNS {
    constructor(client) {
        this._client = client;
    }
    findCorrelation(params, opts) {
        return this._client.call('tools.ozone.signature.findCorrelation', params, undefined, opts);
    }
    findRelatedAccounts(params, opts) {
        return this._client.call('tools.ozone.signature.findRelatedAccounts', params, undefined, opts);
    }
    searchAccounts(params, opts) {
        return this._client.call('tools.ozone.signature.searchAccounts', params, undefined, opts);
    }
}
export class ToolsOzoneTeamNS {
    constructor(client) {
        this._client = client;
    }
    addMember(data, opts) {
        return this._client
            .call('tools.ozone.team.addMember', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneTeamAddMember.toKnownErr(e);
        });
    }
    deleteMember(data, opts) {
        return this._client
            .call('tools.ozone.team.deleteMember', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneTeamDeleteMember.toKnownErr(e);
        });
    }
    listMembers(params, opts) {
        return this._client.call('tools.ozone.team.listMembers', params, undefined, opts);
    }
    updateMember(data, opts) {
        return this._client
            .call('tools.ozone.team.updateMember', opts?.qp, data, opts)
            .catch((e) => {
            throw ToolsOzoneTeamUpdateMember.toKnownErr(e);
        });
    }
}
export class ToolsOzoneVerificationNS {
    constructor(client) {
        this._client = client;
    }
    grantVerifications(data, opts) {
        return this._client.call('tools.ozone.verification.grantVerifications', opts?.qp, data, opts);
    }
    listVerifications(params, opts) {
        return this._client.call('tools.ozone.verification.listVerifications', params, undefined, opts);
    }
    revokeVerifications(data, opts) {
        return this._client.call('tools.ozone.verification.revokeVerifications', opts?.qp, data, opts);
    }
}
