// @atproto/api/dist/client/types/com/atproto/temp/checkHandleAvailability.js
/**
 * GENERATED CODE - DO NOT MODIFY
 */
import { XRPCError } from '@atproto/xrpc';
import { validate as _validate } from '../../../../lexicons.js';
import { is$typed as _is$typed, } from '../../../../util.js';
const is$typed = _is$typed, validate = _validate;
const id = 'com.atproto.temp.checkHandleAvailability';
export class InvalidEmailError extends XRPCError {
    constructor(src) {
        super(src.status, src.error, src.message, src.headers, { cause: src });
    }
}
export function toKnownErr(e) {
    if (e instanceof XRPCError) {
        if (e.error === 'InvalidEmail')
            return new InvalidEmailError(e);
    }
    return e;
}
const hashResultAvailable = 'resultAvailable';
export function isResultAvailable(v) {
    return is$typed(v, id, hashResultAvailable);
}
export function validateResultAvailable(v) {
    return validate(v, id, hashResultAvailable);
}
const hashResultUnavailable = 'resultUnavailable';
export function isResultUnavailable(v) {
    return is$typed(v, id, hashResultUnavailable);
}
export function validateResultUnavailable(v) {
    return validate(v, id, hashResultUnavailable);
}
const hashSuggestion = 'suggestion';
export function isSuggestion(v) {
    return is$typed(v, id, hashSuggestion);
}
export function validateSuggestion(v) {
    return validate(v, id, hashSuggestion);
}
